{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#All-Cell-annotations\" data-toc-modified-id=\"All-Cell-annotations-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>All Cell annotations</a></span></li><li><span><a href=\"#Parse-BALF-GEO-Series\" data-toc-modified-id=\"Parse-BALF-GEO-Series-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Parse BALF GEO Series</a></span><ul class=\"toc-item\"><li><span><a href=\"#Parse-CellRanger-HDF5-files\" data-toc-modified-id=\"Parse-CellRanger-HDF5-files-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Parse CellRanger HDF5 files</a></span><ul class=\"toc-item\"><li><span><a href=\"#File-Format\" data-toc-modified-id=\"File-Format-2.1.1\"><span class=\"toc-item-num\">2.1.1&nbsp;&nbsp;</span>File Format</a></span></li></ul></li><li><span><a href=\"#Parse-CellRanger-3.0+-Feature-Barcode-Matricies\" data-toc-modified-id=\"Parse-CellRanger-3.0+-Feature-Barcode-Matricies-2.2\"><span class=\"toc-item-num\">2.2&nbsp;&nbsp;</span>Parse CellRanger 3.0+ Feature-Barcode Matricies</a></span></li></ul></li><li><span><a href=\"#All-Samples\" data-toc-modified-id=\"All-Samples-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>All Samples</a></span></li><li><span><a href=\"#Get-model-genes\" data-toc-modified-id=\"Get-model-genes-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Get model genes</a></span></li><li><span><a href=\"#Model-genes-for-all-celltypes-per-patient\" data-toc-modified-id=\"Model-genes-for-all-celltypes-per-patient-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Model genes for all celltypes per patient</a></span><ul class=\"toc-item\"><li><span><a href=\"#Write-out-unnormalized-counts-of-model-genes\" data-toc-modified-id=\"Write-out-unnormalized-counts-of-model-genes-5.1\"><span class=\"toc-item-num\">5.1&nbsp;&nbsp;</span>Write out unnormalized counts of model genes</a></span></li><li><span><a href=\"#Library-size-normalization\" data-toc-modified-id=\"Library-size-normalization-5.2\"><span class=\"toc-item-num\">5.2&nbsp;&nbsp;</span>Library size normalization</a></span></li><li><span><a href=\"#Write-out-library-normalized-counts-of-genes\" data-toc-modified-id=\"Write-out-library-normalized-counts-of-genes-5.3\"><span class=\"toc-item-num\">5.3&nbsp;&nbsp;</span>Write out library normalized counts of genes</a></span></li><li><span><a href=\"#All-samples-and-cell-types\" data-toc-modified-id=\"All-samples-and-cell-types-5.4\"><span class=\"toc-item-num\">5.4&nbsp;&nbsp;</span>All samples and cell types</a></span></li><li><span><a href=\"#Extract-samples-for-each-celltype\" data-toc-modified-id=\"Extract-samples-for-each-celltype-5.5\"><span class=\"toc-item-num\">5.5&nbsp;&nbsp;</span>Extract samples for each celltype</a></span></li></ul></li><li><span><a href=\"#Healthy-control-4-(HC4)\" data-toc-modified-id=\"Healthy-control-4-(HC4)-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>Healthy control 4 (HC4)</a></span><ul class=\"toc-item\"><li><span><a href=\"#Parse-Healthy-control-4-(HC4)-GEO-Sample\" data-toc-modified-id=\"Parse-Healthy-control-4-(HC4)-GEO-Sample-6.1\"><span class=\"toc-item-num\">6.1&nbsp;&nbsp;</span>Parse Healthy control 4 (HC4) GEO Sample</a></span></li></ul></li><li><span><a href=\"#Macrophage\" data-toc-modified-id=\"Macrophage-7\"><span class=\"toc-item-num\">7&nbsp;&nbsp;</span>Macrophage</a></span><ul class=\"toc-item\"><li><span><a href=\"#Epithelial-cells\" data-toc-modified-id=\"Epithelial-cells-7.1\"><span class=\"toc-item-num\">7.1&nbsp;&nbsp;</span>Epithelial cells</a></span><ul class=\"toc-item\"><li><span><a href=\"#Severe-COVID-19-patient-2-(S2)-Epithelial-cell-barcodes\" data-toc-modified-id=\"Severe-COVID-19-patient-2-(S2)-Epithelial-cell-barcodes-7.1.1\"><span class=\"toc-item-num\">7.1.1&nbsp;&nbsp;</span>Severe COVID-19 patient 2 (S2) Epithelial cell barcodes</a></span></li><li><span><a href=\"#Healthy-control-4-(HC4)-Epithelial-cell-barcodes\" data-toc-modified-id=\"Healthy-control-4-(HC4)-Epithelial-cell-barcodes-7.1.2\"><span class=\"toc-item-num\">7.1.2&nbsp;&nbsp;</span>Healthy control 4 (HC4) Epithelial cell barcodes</a></span></li></ul></li><li><span><a href=\"#Compare-Raw-HC4-barcodes-with-Metdata-HC4-barcodes\" data-toc-modified-id=\"Compare-Raw-HC4-barcodes-with-Metdata-HC4-barcodes-7.2\"><span class=\"toc-item-num\">7.2&nbsp;&nbsp;</span>Compare Raw HC4 barcodes with Metdata HC4 barcodes</a></span></li></ul></li><li><span><a href=\"#Number-of-Epithelial-cells-in-H4\" data-toc-modified-id=\"Number-of-Epithelial-cells-in-H4-8\"><span class=\"toc-item-num\">8&nbsp;&nbsp;</span>Number of Epithelial cells in H4</a></span><ul class=\"toc-item\"><li><span><a href=\"#Write-out-Healthy-Control-4-epithelial-cells-to-csv\" data-toc-modified-id=\"Write-out-Healthy-Control-4-epithelial-cells-to-csv-8.1\"><span class=\"toc-item-num\">8.1&nbsp;&nbsp;</span>Write out Healthy Control 4 epithelial cells to csv</a></span></li></ul></li><li><span><a href=\"#Severe-COVID-19-patient-2\" data-toc-modified-id=\"Severe-COVID-19-patient-2-9\"><span class=\"toc-item-num\">9&nbsp;&nbsp;</span>Severe COVID-19 patient 2</a></span><ul class=\"toc-item\"><li><span><a href=\"#Parse-Severe-COVID-19-patient-2-(S2)-GEO-sample-file\" data-toc-modified-id=\"Parse-Severe-COVID-19-patient-2-(S2)-GEO-sample-file-9.1\"><span class=\"toc-item-num\">9.1&nbsp;&nbsp;</span>Parse Severe COVID-19 patient 2 (S2) GEO sample file</a></span></li><li><span><a href=\"#Parse-S2-raw-HDF5-file\" data-toc-modified-id=\"Parse-S2-raw-HDF5-file-9.2\"><span class=\"toc-item-num\">9.2&nbsp;&nbsp;</span>Parse S2 raw HDF5 file</a></span></li><li><span><a href=\"#Compare-Raw-S2-barcodes-with-S2-metadata-barcodes\" data-toc-modified-id=\"Compare-Raw-S2-barcodes-with-S2-metadata-barcodes-9.3\"><span class=\"toc-item-num\">9.3&nbsp;&nbsp;</span>Compare Raw S2 barcodes with S2 metadata barcodes</a></span></li><li><span><a href=\"#Healthy-controls\" data-toc-modified-id=\"Healthy-controls-9.4\"><span class=\"toc-item-num\">9.4&nbsp;&nbsp;</span>Healthy controls</a></span></li><li><span><a href=\"#Write-out-epithelial-samples-to-csv\" data-toc-modified-id=\"Write-out-epithelial-samples-to-csv-9.5\"><span class=\"toc-item-num\">9.5&nbsp;&nbsp;</span>Write out epithelial samples to csv</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# All Cell annotations\n",
    "\n",
    "Cell metadata annotations relable the suffix of barcodes [github issue #6](https://github.com/zhangzlab/covid_balf/issues/6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/zuck016/Projects/Proposals/MARS/Pathogenesis/Covid-19-Causal-Reasoning/NeurIPS-determinants-of-viral-pathogenesis\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>celltype</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGACACTAA_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGAGACACTAA_HC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGGAGTACC_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGAGGAGTACC_HC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGAGGATATAC_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGAGGATATAC_HC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGAGGTCATCT_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGAGGTCATCT_HC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGCACGGATAG_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>5</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGCACGGATAG_HC1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   ID sample sample_new group disease hasnCoV  cluster  \\\n",
       "0  AAACCTGAGACACTAA_1    C51        HC1    HC       N       N        3   \n",
       "1  AAACCTGAGGAGTACC_1    C51        HC1    HC       N       N        3   \n",
       "2  AAACCTGAGGATATAC_1    C51        HC1    HC       N       N        3   \n",
       "3  AAACCTGAGGTCATCT_1    C51        HC1    HC       N       N        3   \n",
       "4  AAACCTGCACGGATAG_1    C51        HC1    HC       N       N        5   \n",
       "\n",
       "      celltype               barcode  \n",
       "0  Macrophages  AAACCTGAGACACTAA_HC1  \n",
       "1  Macrophages  AAACCTGAGGAGTACC_HC1  \n",
       "2  Macrophages  AAACCTGAGGATATAC_HC1  \n",
       "3  Macrophages  AAACCTGAGGTCATCT_HC1  \n",
       "4  Macrophages  AAACCTGCACGGATAG_HC1  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd, os\n",
    "import GEOparse\n",
    "all_cells = pd.read_csv('data/all.cell.annotation.meta.txt',sep='\\t')\n",
    "all_cells['barcode'] = all_cells['ID'].str.split('_').str.get(0) + '_' + all_cells['sample_new']\n",
    "all_cells.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "65813"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_barcodes = set(all_cells['barcode'].values.tolist())\n",
    "\n",
    "len(all_barcodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B',\n",
       " 'Epithelial',\n",
       " 'Macrophages',\n",
       " 'Mast',\n",
       " 'NK',\n",
       " 'Neutrophil',\n",
       " 'Plasma',\n",
       " 'T',\n",
       " 'mDC',\n",
       " 'pDC'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "celltypes = set(all_cells['celltype'].values)\n",
    "celltypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>celltype</th>\n",
       "      <th>group</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">B</th>\n",
       "      <th>HC</th>\n",
       "      <td>51</td>\n",
       "      <td>51</td>\n",
       "      <td>51</td>\n",
       "      <td>51</td>\n",
       "      <td>51</td>\n",
       "      <td>51</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Epithelial</th>\n",
       "      <th>HC</th>\n",
       "      <td>1442</td>\n",
       "      <td>1442</td>\n",
       "      <td>1442</td>\n",
       "      <td>1442</td>\n",
       "      <td>1442</td>\n",
       "      <td>1442</td>\n",
       "      <td>1442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>215</td>\n",
       "      <td>215</td>\n",
       "      <td>215</td>\n",
       "      <td>215</td>\n",
       "      <td>215</td>\n",
       "      <td>215</td>\n",
       "      <td>215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>1874</td>\n",
       "      <td>1874</td>\n",
       "      <td>1874</td>\n",
       "      <td>1874</td>\n",
       "      <td>1874</td>\n",
       "      <td>1874</td>\n",
       "      <td>1874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Macrophages</th>\n",
       "      <th>HC</th>\n",
       "      <td>18539</td>\n",
       "      <td>18539</td>\n",
       "      <td>18539</td>\n",
       "      <td>18539</td>\n",
       "      <td>18539</td>\n",
       "      <td>18539</td>\n",
       "      <td>18539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>4353</td>\n",
       "      <td>4353</td>\n",
       "      <td>4353</td>\n",
       "      <td>4353</td>\n",
       "      <td>4353</td>\n",
       "      <td>4353</td>\n",
       "      <td>4353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>26525</td>\n",
       "      <td>26525</td>\n",
       "      <td>26525</td>\n",
       "      <td>26525</td>\n",
       "      <td>26525</td>\n",
       "      <td>26525</td>\n",
       "      <td>26525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Mast</th>\n",
       "      <th>HC</th>\n",
       "      <td>25</td>\n",
       "      <td>25</td>\n",
       "      <td>25</td>\n",
       "      <td>25</td>\n",
       "      <td>25</td>\n",
       "      <td>25</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">NK</th>\n",
       "      <th>HC</th>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>298</td>\n",
       "      <td>298</td>\n",
       "      <td>298</td>\n",
       "      <td>298</td>\n",
       "      <td>298</td>\n",
       "      <td>298</td>\n",
       "      <td>298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>686</td>\n",
       "      <td>686</td>\n",
       "      <td>686</td>\n",
       "      <td>686</td>\n",
       "      <td>686</td>\n",
       "      <td>686</td>\n",
       "      <td>686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Neutrophil</th>\n",
       "      <th>M</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>1603</td>\n",
       "      <td>1603</td>\n",
       "      <td>1603</td>\n",
       "      <td>1603</td>\n",
       "      <td>1603</td>\n",
       "      <td>1603</td>\n",
       "      <td>1603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Plasma</th>\n",
       "      <th>HC</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>1024</td>\n",
       "      <td>1024</td>\n",
       "      <td>1024</td>\n",
       "      <td>1024</td>\n",
       "      <td>1024</td>\n",
       "      <td>1024</td>\n",
       "      <td>1024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">T</th>\n",
       "      <th>HC</th>\n",
       "      <td>1225</td>\n",
       "      <td>1225</td>\n",
       "      <td>1225</td>\n",
       "      <td>1225</td>\n",
       "      <td>1225</td>\n",
       "      <td>1225</td>\n",
       "      <td>1225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>2135</td>\n",
       "      <td>2135</td>\n",
       "      <td>2135</td>\n",
       "      <td>2135</td>\n",
       "      <td>2135</td>\n",
       "      <td>2135</td>\n",
       "      <td>2135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>4356</td>\n",
       "      <td>4356</td>\n",
       "      <td>4356</td>\n",
       "      <td>4356</td>\n",
       "      <td>4356</td>\n",
       "      <td>4356</td>\n",
       "      <td>4356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">mDC</th>\n",
       "      <th>HC</th>\n",
       "      <td>485</td>\n",
       "      <td>485</td>\n",
       "      <td>485</td>\n",
       "      <td>485</td>\n",
       "      <td>485</td>\n",
       "      <td>485</td>\n",
       "      <td>485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>159</td>\n",
       "      <td>159</td>\n",
       "      <td>159</td>\n",
       "      <td>159</td>\n",
       "      <td>159</td>\n",
       "      <td>159</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>334</td>\n",
       "      <td>334</td>\n",
       "      <td>334</td>\n",
       "      <td>334</td>\n",
       "      <td>334</td>\n",
       "      <td>334</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">pDC</th>\n",
       "      <th>HC</th>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>76</td>\n",
       "      <td>76</td>\n",
       "      <td>76</td>\n",
       "      <td>76</td>\n",
       "      <td>76</td>\n",
       "      <td>76</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ID  sample  sample_new  disease  hasnCoV  cluster  \\\n",
       "celltype    group                                                         \n",
       "B           HC        51      51          51       51       51       51   \n",
       "            M         67      67          67       67       67       67   \n",
       "            S        102     102         102      102      102      102   \n",
       "Epithelial  HC      1442    1442        1442     1442     1442     1442   \n",
       "            M        215     215         215      215      215      215   \n",
       "            S       1874    1874        1874     1874     1874     1874   \n",
       "Macrophages HC     18539   18539       18539    18539    18539    18539   \n",
       "            M       4353    4353        4353     4353     4353     4353   \n",
       "            S      26525   26525       26525    26525    26525    26525   \n",
       "Mast        HC        25      25          25       25       25       25   \n",
       "            M          6       6           6        6        6        6   \n",
       "            S         39      39          39       39       39       39   \n",
       "NK          HC        97      97          97       97       97       97   \n",
       "            M        298     298         298      298      298      298   \n",
       "            S        686     686         686      686      686      686   \n",
       "Neutrophil  M          4       4           4        4        4        4   \n",
       "            S       1603    1603        1603     1603     1603     1603   \n",
       "Plasma      HC         3       3           3        3        3        3   \n",
       "            M         14      14          14       14       14       14   \n",
       "            S       1024    1024        1024     1024     1024     1024   \n",
       "T           HC      1225    1225        1225     1225     1225     1225   \n",
       "            M       2135    2135        2135     2135     2135     2135   \n",
       "            S       4356    4356        4356     4356     4356     4356   \n",
       "mDC         HC       485     485         485      485      485      485   \n",
       "            M        159     159         159      159      159      159   \n",
       "            S        334     334         334      334      334      334   \n",
       "pDC         HC        16      16          16       16       16       16   \n",
       "            M         60      60          60       60       60       60   \n",
       "            S         76      76          76       76       76       76   \n",
       "\n",
       "                   barcode  \n",
       "celltype    group           \n",
       "B           HC          51  \n",
       "            M           67  \n",
       "            S          102  \n",
       "Epithelial  HC        1442  \n",
       "            M          215  \n",
       "            S         1874  \n",
       "Macrophages HC       18539  \n",
       "            M         4353  \n",
       "            S        26525  \n",
       "Mast        HC          25  \n",
       "            M            6  \n",
       "            S           39  \n",
       "NK          HC          97  \n",
       "            M          298  \n",
       "            S          686  \n",
       "Neutrophil  M            4  \n",
       "            S         1603  \n",
       "Plasma      HC           3  \n",
       "            M           14  \n",
       "            S         1024  \n",
       "T           HC        1225  \n",
       "            M         2135  \n",
       "            S         4356  \n",
       "mDC         HC         485  \n",
       "            M          159  \n",
       "            S          334  \n",
       "pDC         HC          16  \n",
       "            M           60  \n",
       "            S           76  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_counts = all_cells.groupby(['celltype', 'group']).count()\n",
    "cell_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['HC1',\n",
       " 'HC2',\n",
       " 'HC3',\n",
       " 'HC4',\n",
       " 'M1',\n",
       " 'M2',\n",
       " 'M3',\n",
       " 'S1',\n",
       " 'S2',\n",
       " 'S3',\n",
       " 'S4',\n",
       " 'S5',\n",
       " 'S6']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patients = sorted(set(all_cells['sample_new'].values))\n",
    "patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'HC1': 'C51',\n",
       " 'HC2': 'C52',\n",
       " 'HC3': 'C100',\n",
       " 'HC4': 'GSM3660650',\n",
       " 'M1': 'C141',\n",
       " 'M2': 'C142',\n",
       " 'M3': 'C144',\n",
       " 'S1': 'C145',\n",
       " 'S2': 'C143',\n",
       " 'S3': 'C146',\n",
       " 'S4': 'C148',\n",
       " 'S5': 'C149',\n",
       " 'S6': 'C152'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient_to_sample = dict(all_cells.groupby(['sample_new', 'sample']).count().index.tolist())\n",
    "patient_to_sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'C100': 'HC3',\n",
       " 'C141': 'M1',\n",
       " 'C142': 'M2',\n",
       " 'C143': 'S2',\n",
       " 'C144': 'M3',\n",
       " 'C145': 'S1',\n",
       " 'C146': 'S3',\n",
       " 'C148': 'S4',\n",
       " 'C149': 'S5',\n",
       " 'C152': 'S6',\n",
       " 'C51': 'HC1',\n",
       " 'C52': 'HC2',\n",
       " 'GSM3660650': 'HC4'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_to_patient = dict(all_cells.groupby(['sample','sample_new']).count().index.tolist())\n",
    "sample_to_patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_celltype_barcodes_from_patient( patient,  celltype, all_cells):\n",
    "    return set(all_cells[(all_cells['sample_new'] == patient) &\n",
    "                                               (all_cells['celltype'] == celltype)]['barcode']\n",
    "                                    .values.tolist())\n",
    "def extract_celltype_from_patient(patient, raw, celltype, all_cells ):\n",
    "    old_barcodes = raw[patient].columns\n",
    "    raw[patient].columns = [bc.split('-')[0] for bc in old_barcodes]\n",
    "    celltype_barcodes = get_celltype_barcodes_from_patient( patient, raw, celltype, all_cells)\n",
    "    return raw[patient][sorted(celltype_barcodes)]\n",
    "\n",
    "def get_metadata_barcodes_from_patient(patient, raw, all_cells):\n",
    "    return set(all_cells[all_cells['sample_new'] == patient]['ID']\n",
    "                                    .str.split('_').str.get(0)\n",
    "                                    .values.tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parse BALF GEO Series"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GSM4339769_C141_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339770_C142_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339771_C143_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339772_C144_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339773_C145_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4339774_C146_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475048_C51_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475049_C52_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475050_C100_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475051_C148_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475052_C149_filtered_feature_bc_matrix.h5',\n",
       " 'GSM4475053_C152_filtered_feature_bc_matrix.h5']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os \n",
    "import GEOparse\n",
    "mars_dir = '../../MARS-Pathogenesis/'\n",
    "balf_dir = os.path.join(mars_dir,'Coronavirus/SARS-CoV-2/COVID-BALF/')\n",
    "balf_files = [f for f in os.listdir(balf_dir) if '.h5' in f]\n",
    "balf_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "13-Jun-2020 15:24:42 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "13-Jun-2020 15:24:42 INFO GEOparse - File already exist: using local version.\n",
      "13-Jun-2020 15:24:42 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSE145926_family.soft.gz: \n",
      "13-Jun-2020 15:24:42 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "13-Jun-2020 15:24:42 DEBUG GEOparse - SERIES: GSE145926\n",
      "13-Jun-2020 15:24:42 DEBUG GEOparse - PLATFORM: GPL23227\n",
      "13-Jun-2020 15:24:42 DEBUG GEOparse - SAMPLE: GSM4339769\n",
      "13-Jun-2020 15:24:42 DEBUG GEOparse - SAMPLE: GSM4339770\n",
      "13-Jun-2020 15:24:42 DEBUG GEOparse - SAMPLE: GSM4339771\n",
      "13-Jun-2020 15:24:42 DEBUG GEOparse - SAMPLE: GSM4339772\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4339773\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4339774\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4385990\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4385991\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4385992\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4385993\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4385994\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4385995\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4475048\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4475049\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4475050\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4475051\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4475052\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4475053\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4475054\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4475055\n",
      "13-Jun-2020 15:24:43 DEBUG GEOparse - SAMPLE: GSM4475056\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name:  GSM4339769\n",
      "Metadata:\n",
      " - title : BALF, C141 (scRNA-seq)\n",
      " - geo_accession : GSM4339769\n",
      " - status : Public on Apr 22 2020\n",
      " - submission_date : Feb 25 2020\n",
      " - last_update_date : Apr 22 2020\n",
      " - type : SRA\n",
      " - channel_count : 1\n",
      " - source_name_ch1 : BALF\n",
      " - organism_ch1 : Homo sapiens\n",
      " - taxid_ch1 : 9606\n",
      " - characteristics_ch1 : disease state: Coronavirus infected disease-19 (COVID-19), patient group: mild, tissue: lung, sample type: BALF, cell subsets: Total cell\n",
      " - molecule_ch1 : polyA RNA\n",
      " - extract_protocol_ch1 : Total 11 µl of single cell suspension and 40 µl barcoded Gel Beads were loaded to Chromium Chip A to generate single-cell gel bead-in-emulsion (GEM). The poly-adenylated transcripts were reverse-transcribed later. The single-cell capturing and downstream library constructions were performed using the Chromium Single Cell 5’ library preparation kit according to the manufacturer’s protocol (10x Genomics). Full-length cDNA along with cell-barcode identifiers were PCR-amplified and sequencing libraries were prepared and normalized to 3 nM.\n",
      " - description : Single-cell RNAseq data\n",
      " - data_processing : The Cell Ranger Software Suite (Version 3.1.0) was used to perform sample de-multiplexing, barcode processing and single-cell 5’ UMI counting with human GRCh38 as the reference genome. Specifically, splicing-aware aligner STAR was used in FASTQs alignment. Cell barcodes were then determined based on distribution of UMI count automatically. Finally, gene-barcode matrix of all 6 donors and 8 previously reported healthy control was integrated with Seurat v3 to remove batch effect across different donors. Following criteria were then applied to each cell, i.e., gene number between 200 and 6000, UMI count above 1000 and mitochondrial gene percentage below 0.1., The filtered gene-barcode matrix was normalized with LogNormalize methods in Seurat and analyzed by principal component analysis (PCA) using the top 2, 000 most variable genes. Then Uniform Manifold Approximation and Projection (UMAP) was performed on the top 50 principal components for visualizing the cells. Meanwhile, graph-based clustering was performed on the PCA-reduced data for clustering analysis with Seurat v3., MAST in Seurat v3 was used to perform differential analysis. For each cluster, differentially-expressed genes (DEGs) were generated relative to all of the other cells., The TCR sequences for each single T cell were assembled by Cell Ranger vdj pipeline (v3.1.0), leading to the identification of CDR3 sequence and the rearranged TCR gene. Cells with both TCR alpha and beta chains were kept and cells with only one TCR chain were discarded., Genome_build: GRCh38 + COVID(MN908947), Supplementary_files_format_and_content: h5, raw count matrix generated by Cell Ranger count; csv, tcr contig annotations generated by Cell Ranger vdj\n",
      " - platform_id : GPL23227\n",
      " - contact_name : Zheng,,Zhang\n",
      " - contact_institute : Shenzhen 3rd People's Hospital\n",
      " - contact_address : No. 29, Bulan Road\n",
      " - contact_city : Shenzhen\n",
      " - contact_state : Guangdong\n",
      " - contact_zip/postal_code : 454171\n",
      " - contact_country : China\n",
      " - instrument_model : BGISEQ-500\n",
      " - library_selection : cDNA\n",
      " - library_source : transcriptomic\n",
      " - library_strategy : RNA-Seq\n",
      " - relation : BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN14207961, SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX7802546\n",
      " - supplementary_file_1 : ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4339nnn/GSM4339769/suppl/GSM4339769_C141_filtered_feature_bc_matrix.h5\n",
      " - series_id : GSE145926\n",
      " - data_row_count : 0\n",
      "Table data:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "\n",
    "gse = GEOparse.get_GEO(geo=\"GSE145926\", destdir=balf_dir)\n",
    "\n",
    "for gsm_name, gsm in gse.gsms.items():\n",
    "    print(\"Name: \", gsm_name)\n",
    "    print(\"Metadata:\",)\n",
    "    for key, value in gsm.metadata.items():\n",
    "        print(\" - %s : %s\" % (key, \", \".join(value)))\n",
    "    print (\"Table data:\",)\n",
    "    print (gsm.table.head())\n",
    "    break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse CellRanger HDF5 files\n",
    "\n",
    "Convert [PyTables functions as recommended by 10xGenomics](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices) into a dataframe\n",
    "\n",
    "### File Format\n",
    "\n",
    "The top level of the file contains a single HDF5 group, called matrix, and metadata stored as HDF5 attributes. Within the matrix group are datasets containing the dimensions of the matrix, the matrix entries, as well as the features and cell-barcodes associated with the matrix rows and columns, respectively.\n",
    "\n",
    "|Column\t|Type\t|Description|\n",
    "|-------|-------|----------|\n",
    "|barcodes|\t`string`|\tBarcode sequences and their corresponding GEM wells (e.g. AAACGGGCAGCTCGAC-1)\n",
    "data|\t`uint32`|\tNonzero UMI counts in column-major order\n",
    "indices|\t`uint32`|\tZero-based row index of corresponding element in data\n",
    "indptr|\t`uint32`\t|Zero-based index into data / indices of the start of each column, i.e., the data corresponding to each barcode sequence\n",
    "shape\t|`uint64`\t|Tuple of (# rows, # columns) indicating the matrix dimensions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import collections\n",
    "import scipy.sparse as sp_sparse\n",
    "import tables\n",
    " \n",
    "CountMatrix = collections.namedtuple('CountMatrix', ['feature_ref', 'barcodes', 'matrix'])\n",
    "\n",
    "def b2a(fg, attr):\n",
    "    return [gene.decode(\"utf-8\")  for gene in getattr(fg, attr).read().tolist()]\n",
    "def get_matrix_from_h5(filename):\n",
    "    with tables.open_file(filename, 'r') as f:\n",
    "        mat_group = f.get_node(f.root, 'matrix')\n",
    "        barcodes = [barcode.decode('utf-8') for barcode in f.get_node(mat_group, 'barcodes').read().tolist()]\n",
    "        data = getattr(mat_group, 'data').read()\n",
    "        indices = getattr(mat_group, 'indices').read()\n",
    "        indptr = getattr(mat_group, 'indptr').read()\n",
    "        shape = getattr(mat_group, 'shape').read()\n",
    "        matrix = sp_sparse.csc_matrix((data, indices, indptr), shape=shape)\n",
    "         \n",
    "        feature_ref = {}\n",
    "        feature_group = f.get_node(mat_group, 'features')\n",
    "        feature_ids = b2a(feature_group, 'id')\n",
    "        feature_names = b2a(feature_group, 'name')\n",
    "        feature_types = b2a(feature_group, 'feature_type')\n",
    "        feature_ref['id'] = feature_ids\n",
    "        feature_ref['name'] = feature_names\n",
    "        feature_ref['feature_type'] = feature_types\n",
    "        tag_keys = b2a(feature_group, '_all_tag_keys')\n",
    "        for key in tag_keys:\n",
    "            if type(key) is not str:\n",
    "                key = key.decode(\"utf-8\") \n",
    "                feature_ref[key] = b2a(feature_group, key)\n",
    "            else:\n",
    "                print(type(key), key)\n",
    "        cmx = CountMatrix(feature_ref, barcodes, matrix)\n",
    "        gene_names= cmx.feature_ref['name']\n",
    "        barcodes =  cmx.barcodes\n",
    "        mat = cmx.matrix\n",
    "        rows,cols = mat.shape\n",
    "        if len(gene_names) == rows and len(barcodes) == cols:  \n",
    "            return pd.DataFrame.sparse.from_spmatrix(data=mat, columns=barcodes, index=gene_names)\n",
    "        elif len(gene_names) == cols and len(barcodes) == rows:  \n",
    "            return pd.DataFrame.sparse.from_spmatrix(data=mat, index=barcodes, columns=gene_names )\n",
    "        else:\n",
    "            print(\"Rows, cols =\", rows, cols, \"genes, barcodes=\",len(gene_names), len(barcodes))\n",
    "    \n",
    " \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse CellRanger 3.0+ Feature-Barcode Matricies\n",
    "\n",
    "[CellRanger Format description according to 10xgenomics](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices):\n",
    "\n",
    "Each matrix is stored in the [Market Exchange Format (MEX)](https://math.nist.gov/MatrixMarket/formats.html) for sparse matrices. It also contains gzipped TSV files with feature and barcode sequences corresponding to row and column indices respectively. For example, the matrices output may look like\n",
    "```\n",
    "$ cd /home/jdoe/runs/sample345/outs\n",
    "$ tree filtered_feature_bc_matrix\n",
    "filtered_feature_bc_matrix\n",
    "├── barcodes.tsv.gz\n",
    "├── features.tsv.gz\n",
    "└── matrix.mtx.gz\n",
    "0 directories, 3 files\n",
    "```\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip, csv\n",
    "import os\n",
    "import scipy.io\n",
    " \n",
    "\n",
    "def parse_mtx(barcodes_path, features_path, mtx_path):\n",
    "    \"\"\"Return a dataframe with barcodes and features and matrix data\"\"\"\n",
    "    mat = scipy.io.mmread(mtx_path)\n",
    "    rows, cols = mat.shape\n",
    "    features = pd.read_csv(features_path,delimiter='\\t',compression='gzip',header=None,\n",
    "                           names=['feature_id','gene_name', 'feature_type'])\n",
    "    \n",
    "    barcodes = pd.read_csv(barcodes_path, delimiter=\"\\t\", \n",
    "                           compression='gzip', \n",
    "                           header=None,\n",
    "                           names=['barcode']\n",
    "                           )\n",
    "    if len(features.index) == rows and len(barcodes.index) == cols:  \n",
    "        return pd.DataFrame.sparse.from_spmatrix(data=mat, columns=barcodes['barcode'], index=features['gene_name'])\n",
    "    elif len(features.index) == cols and len(barcodes.index) == rows:  \n",
    "        return pd.DataFrame.sparse.from_spmatrix(data=mat, index=barcodes['barcode'], columns=features['gene_name'] )\n",
    "    else:\n",
    "        print(\"Rows, cols =\", rows, cols, \"genes, barcodes=\",len(gene_names), len(barcodes))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# All Samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['GSM4339769_C141_filtered_feature_bc_matrix.h5',\n",
       "  'GSM4339770_C142_filtered_feature_bc_matrix.h5',\n",
       "  'GSM4339771_C143_filtered_feature_bc_matrix.h5',\n",
       "  'GSM4339772_C144_filtered_feature_bc_matrix.h5',\n",
       "  'GSM4339773_C145_filtered_feature_bc_matrix.h5',\n",
       "  'GSM4339774_C146_filtered_feature_bc_matrix.h5',\n",
       "  'GSM4475048_C51_filtered_feature_bc_matrix.h5',\n",
       "  'GSM4475049_C52_filtered_feature_bc_matrix.h5',\n",
       "  'GSM4475050_C100_filtered_feature_bc_matrix.h5',\n",
       "  'GSM4475051_C148_filtered_feature_bc_matrix.h5',\n",
       "  'GSM4475052_C149_filtered_feature_bc_matrix.h5',\n",
       "  'GSM4475053_C152_filtered_feature_bc_matrix.h5'],\n",
       " {'C100': 'HC3',\n",
       "  'C141': 'M1',\n",
       "  'C142': 'M2',\n",
       "  'C143': 'S2',\n",
       "  'C144': 'M3',\n",
       "  'C145': 'S1',\n",
       "  'C146': 'S3',\n",
       "  'C148': 'S4',\n",
       "  'C149': 'S5',\n",
       "  'C152': 'S6',\n",
       "  'C51': 'HC1',\n",
       "  'C52': 'HC2',\n",
       "  'GSM3660650': 'HC4'})"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "balf_files, sample_to_patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_raw_barcodes_for_patient(raw_patient):\n",
    "    return set([raw_barcode.split('-')[0] for raw_barcode in raw_patient.columns])\n",
    "def get_all_samples(balf_files, sample_to_patient ): \n",
    "    raw = {}\n",
    "    geo = {}\n",
    "    raw_barcodes = {}\n",
    "    for gsm_file in balf_files:\n",
    "        gsm_id, sample = gsm_file.split('_')[0:2]\n",
    "        patient = sample_to_patient[sample]\n",
    "        geo[patient] = GEOparse.get_GEO(geo=gsm_id, destdir=balf_dir)\n",
    "        if patient == 'HC4':\n",
    "            raw[patient] = parse_mtx(os.path.join(balf_dir,'GSM3660650_SC249NORbal_fresh_barcodes.tsv.gz'),\n",
    "                                     os.path.join(balf_dir,'GSM3660650_SC249NORbal_fresh_genes.tsv.gz'),\n",
    "                                     os.path.join(balf_dir,'GSM3660650_SC249NORbal_fresh_matrix.mtx.gz'))\n",
    "        else:\n",
    "            raw[patient] = get_matrix_from_h5(os.path.join(balf_dir,gsm_file))\n",
    "        raw[patient].columns = [bc.split('-')[0] + '_' + patient for bc in raw[patient].columns]\n",
    "        raw_barcodes[patient] = raw[patient].columns\n",
    "    return geo, raw, raw_barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14-Jun-2020 13:12:18 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "14-Jun-2020 13:12:18 INFO GEOparse - File already exist: using local version.\n",
      "14-Jun-2020 13:12:18 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM3660650.txt: \n",
      "14-Jun-2020 13:17:55 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "14-Jun-2020 13:17:55 INFO GEOparse - File already exist: using local version.\n",
      "14-Jun-2020 13:17:55 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4339769.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'> genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14-Jun-2020 13:17:59 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "14-Jun-2020 13:17:59 INFO GEOparse - File already exist: using local version.\n",
      "14-Jun-2020 13:17:59 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4339770.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'> genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14-Jun-2020 13:18:05 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "14-Jun-2020 13:18:05 INFO GEOparse - File already exist: using local version.\n",
      "14-Jun-2020 13:18:05 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4339771.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'> genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14-Jun-2020 13:18:19 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "14-Jun-2020 13:18:19 INFO GEOparse - File already exist: using local version.\n",
      "14-Jun-2020 13:18:19 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4339772.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'> genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14-Jun-2020 13:18:21 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "14-Jun-2020 13:18:21 INFO GEOparse - File already exist: using local version.\n",
      "14-Jun-2020 13:18:21 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4339773.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'> genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14-Jun-2020 13:18:30 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "14-Jun-2020 13:18:30 INFO GEOparse - File already exist: using local version.\n",
      "14-Jun-2020 13:18:30 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4339774.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'> genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14-Jun-2020 13:18:33 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "14-Jun-2020 13:18:33 INFO GEOparse - File already exist: using local version.\n",
      "14-Jun-2020 13:18:33 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4475048.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'> genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14-Jun-2020 13:18:39 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "14-Jun-2020 13:18:39 INFO GEOparse - File already exist: using local version.\n",
      "14-Jun-2020 13:18:39 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4475049.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'> genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14-Jun-2020 13:18:45 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "14-Jun-2020 13:18:45 INFO GEOparse - File already exist: using local version.\n",
      "14-Jun-2020 13:18:45 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4475050.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'> genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14-Jun-2020 13:18:50 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "14-Jun-2020 13:18:50 INFO GEOparse - File already exist: using local version.\n",
      "14-Jun-2020 13:18:50 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4475051.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'> genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14-Jun-2020 13:18:53 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "14-Jun-2020 13:18:53 INFO GEOparse - File already exist: using local version.\n",
      "14-Jun-2020 13:18:53 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4475052.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'> genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "14-Jun-2020 13:18:54 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "14-Jun-2020 13:18:54 INFO GEOparse - File already exist: using local version.\n",
      "14-Jun-2020 13:18:54 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4475053.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'> genome\n"
     ]
    }
   ],
   "source": [
    "geo, raw, raw_barcodes = get_all_samples( ['GSM3660650_GSM3660650'] +balf_files, sample_to_patient )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get model genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Temporary Network v_akJ3f4e--_\n",
      "Number of Nodes: 16\n",
      "Number of Edges: 21\n",
      "Number of Citations: 1\n",
      "Number of Authors: 0\n",
      "Network Density: 8.75E-02\n",
      "Number of Components: 1\n",
      "Number of Warnings: 0\n"
     ]
    }
   ],
   "source": [
    "import pybel, os\n",
    "il6_dir = '../NeurIPS-determinants-of-viral-pathogenesis/BEL'\n",
    "il6_amp_file = 'COVID-19-a-new-virus-but-an-old-cytokine-release-syndrome.bel.nodelink.json'\n",
    "il6_amp = pybel.from_nodelink_file(os.path.join(il6_dir, il6_amp_file))\n",
    "il6_amp.summarize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IL6\n",
      "STAT3\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['IL6',\n",
       " 'STAT3',\n",
       " 'ACE2',\n",
       " 'ADAM17',\n",
       " 'AGTR1',\n",
       " 'EGF',\n",
       " 'IL6R',\n",
       " 'TNF',\n",
       " 'EGFR',\n",
       " 'NFKB1',\n",
       " 'FCN1']"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_model_genes( model ):\n",
    "    model_genes = []\n",
    "    for node in il6_amp.nodes:\n",
    "        if  not isinstance(node, pybel.dsl.ComplexAbundance):\n",
    "            if node.namespace == 'HGNC':\n",
    "                model_genes.append(node.name)\n",
    "        else:\n",
    "            print('\\n'.join([m.name for m in node.members]))\n",
    "    return model_genes\n",
    "\n",
    "model_genes = get_model_genes( il6_amp )+ ['TNF', \"EGFR\", \"NFKB1\", \"FCN1\"]\n",
    "model_genes "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Model genes for all celltypes per patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "patient='HC4'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>celltype</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>62920</th>\n",
       "      <td>AAACCTGGTCAGTGGA_13</td>\n",
       "      <td>C152</td>\n",
       "      <td>S6</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>6</td>\n",
       "      <td>T</td>\n",
       "      <td>AAACCTGGTCAGTGGA_S6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62924</th>\n",
       "      <td>AAACGGGGTTACCAGT_13</td>\n",
       "      <td>C152</td>\n",
       "      <td>S6</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>9</td>\n",
       "      <td>T</td>\n",
       "      <td>AAACGGGGTTACCAGT_S6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62934</th>\n",
       "      <td>AAAGCAAGTGATAAGT_13</td>\n",
       "      <td>C152</td>\n",
       "      <td>S6</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>9</td>\n",
       "      <td>T</td>\n",
       "      <td>AAAGCAAGTGATAAGT_S6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62935</th>\n",
       "      <td>AAAGCAAGTTTCGCTC_13</td>\n",
       "      <td>C152</td>\n",
       "      <td>S6</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>14</td>\n",
       "      <td>T</td>\n",
       "      <td>AAAGCAAGTTTCGCTC_S6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62946</th>\n",
       "      <td>AAATGCCTCCGCATCT_13</td>\n",
       "      <td>C152</td>\n",
       "      <td>S6</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>6</td>\n",
       "      <td>T</td>\n",
       "      <td>AAATGCCTCCGCATCT_S6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65791</th>\n",
       "      <td>TTTCCTCCAGGTCTCG_13</td>\n",
       "      <td>C152</td>\n",
       "      <td>S6</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>14</td>\n",
       "      <td>T</td>\n",
       "      <td>TTTCCTCCAGGTCTCG_S6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65798</th>\n",
       "      <td>TTTGCGCTCTAACCGA_13</td>\n",
       "      <td>C152</td>\n",
       "      <td>S6</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>9</td>\n",
       "      <td>T</td>\n",
       "      <td>TTTGCGCTCTAACCGA_S6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65799</th>\n",
       "      <td>TTTGCGCTCTGCTTGC_13</td>\n",
       "      <td>C152</td>\n",
       "      <td>S6</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>14</td>\n",
       "      <td>T</td>\n",
       "      <td>TTTGCGCTCTGCTTGC_S6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65804</th>\n",
       "      <td>TTTGGTTCACTATCTT_13</td>\n",
       "      <td>C152</td>\n",
       "      <td>S6</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>14</td>\n",
       "      <td>T</td>\n",
       "      <td>TTTGGTTCACTATCTT_S6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65809</th>\n",
       "      <td>TTTGTCAAGAAACGCC_13</td>\n",
       "      <td>C152</td>\n",
       "      <td>S6</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>14</td>\n",
       "      <td>T</td>\n",
       "      <td>TTTGTCAAGAAACGCC_S6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>526 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        ID sample sample_new group disease hasnCoV  cluster  \\\n",
       "62920  AAACCTGGTCAGTGGA_13   C152         S6     S       Y       N        6   \n",
       "62924  AAACGGGGTTACCAGT_13   C152         S6     S       Y       N        9   \n",
       "62934  AAAGCAAGTGATAAGT_13   C152         S6     S       Y       N        9   \n",
       "62935  AAAGCAAGTTTCGCTC_13   C152         S6     S       Y       N       14   \n",
       "62946  AAATGCCTCCGCATCT_13   C152         S6     S       Y       N        6   \n",
       "...                    ...    ...        ...   ...     ...     ...      ...   \n",
       "65791  TTTCCTCCAGGTCTCG_13   C152         S6     S       Y       N       14   \n",
       "65798  TTTGCGCTCTAACCGA_13   C152         S6     S       Y       N        9   \n",
       "65799  TTTGCGCTCTGCTTGC_13   C152         S6     S       Y       N       14   \n",
       "65804  TTTGGTTCACTATCTT_13   C152         S6     S       Y       N       14   \n",
       "65809  TTTGTCAAGAAACGCC_13   C152         S6     S       Y       N       14   \n",
       "\n",
       "      celltype              barcode  \n",
       "62920        T  AAACCTGGTCAGTGGA_S6  \n",
       "62924        T  AAACGGGGTTACCAGT_S6  \n",
       "62934        T  AAAGCAAGTGATAAGT_S6  \n",
       "62935        T  AAAGCAAGTTTCGCTC_S6  \n",
       "62946        T  AAATGCCTCCGCATCT_S6  \n",
       "...        ...                  ...  \n",
       "65791        T  TTTCCTCCAGGTCTCG_S6  \n",
       "65798        T  TTTGCGCTCTAACCGA_S6  \n",
       "65799        T  TTTGCGCTCTGCTTGC_S6  \n",
       "65804        T  TTTGGTTCACTATCTT_S6  \n",
       "65809        T  TTTGTCAAGAAACGCC_S6  \n",
       "\n",
       "[526 rows x 9 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_cells[(all_cells['sample_new'] == patient)& (all_cells['celltype'] == celltype)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write out unnormalized counts of model genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "for patient in patients:\n",
    "    for celltype in celltypes:\n",
    "        patient_celltype_barcodes= get_celltype_barcodes_from_patient( patient,  celltype, all_cells)\n",
    "        patient_celltype_genes = raw[patient].loc[model_genes,patient_celltype_barcodes].T\n",
    "        celltype_dir = os.path.join(balf_dir, celltype )\n",
    "\n",
    "        if not os.path.exists(celltype_dir):\n",
    "            os.mkdir(celltype_dir)\n",
    "        patient_celltype_genes.to_csv(\n",
    "            os.path.join(celltype_dir,\n",
    "                         '{}_{}_model_genes.csv'.format(patient, celltype)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Library size normalization\n",
    "\n",
    "Library size normalization is the simplest strategy for performing scaling normalization. We define the library size as the total sum of counts across all genes for each cell, the expected value of which is assumed to scale with any cell-specific biases. The “library size factor” for each cell is then directly proportional to its library size where the proportionality constant is defined such that the mean size factor across all cells is equal to 1. This definition ensures that the normalized expression values are on the same scale as the original counts, which is useful for interpretation - especially when dealing with transformed data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def library_size_factors( df, celltypes, patients ):\n",
    "    library_size = {}\n",
    "    library_sum = 0\n",
    "    for patient in patients:\n",
    "        for celltype in celltypes:\n",
    "            patient_celltype_barcodes= get_celltype_barcodes_from_patient( patient,  celltype, all_cells)\n",
    "            library_size[patient,celltype] = raw[patient][patient_celltype_barcodes].sum(axis=1)\n",
    "            library_sum += library_size[patient,celltype].sum()\n",
    "    library_size_factor = {}\n",
    "    for patient in patients:\n",
    "        for celltype in celltypes:\n",
    "            library_size_factor[patient,celltype] = library_size[patient,celltype]/library_sum\n",
    "    return library_size_factor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "lsf = library_size_factors( raw, celltypes, patients )     \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "def log_norm_counts( df, size_factors, pseudocounts=1, base=2):\n",
    "    if base==2:\n",
    "        return np.log2(df + pseudocounts).subtract(np.log2(size_factors + pseudocounts), axis='index')\n",
    "    elif base:\n",
    "        return np.log(df + pseudocounts).subtract(np.log(size_factors + pseudocounts) -np.log(base), axis='index')\n",
    "    else:\n",
    "        return np.log(df + pseudocounts).subtract(np.log(size_factors + pseudocounts), axis='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IL6       0.000000e+00\n",
       "STAT3     1.611377e-06\n",
       "ACE2      0.000000e+00\n",
       "ADAM17    8.340733e-07\n",
       "AGTR1     0.000000e+00\n",
       "EGF       0.000000e+00\n",
       "IL6R      5.895283e-08\n",
       "TNF       5.676939e-08\n",
       "EGFR      0.000000e+00\n",
       "NFKB1     1.423602e-06\n",
       "FCN1      2.620126e-08\n",
       "dtype: float64"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lsf[patient, celltype][model_genes] "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write out library normalized counts of genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np \n",
    "library_normalized_patient_celltype_genes = {}\n",
    "for patient in patients:\n",
    "    for celltype in celltypes:\n",
    "        patient_celltype_barcodes= get_celltype_barcodes_from_patient( patient,  celltype, all_cells)\n",
    "        library_normalized_patient_celltype_genes[patient, celltype] = log_norm_counts(\n",
    "            raw[patient].loc[model_genes,patient_celltype_barcodes],\n",
    "            lsf[patient,celltype][model_genes])\n",
    "        celltype_dir = os.path.join(balf_dir, celltype )\n",
    "\n",
    "        if not os.path.exists(celltype_dir):\n",
    "            os.mkdir(celltype_dir)\n",
    "        library_normalized_patient_celltype_genes[patient,celltype].T.to_csv(\n",
    "            os.path.join(celltype_dir,\n",
    "                         'log_norm_count_{}_{}_model_genes.csv'.format(patient, celltype)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ATTGGACCAGGGATTG_S6</th>\n",
       "      <th>CATCCACGTCATGCAT_S6</th>\n",
       "      <th>ACCTTTAAGAACAATC_S6</th>\n",
       "      <th>CCTTTCTGTAGGCATG_S6</th>\n",
       "      <th>TATTACCCAAAGTGCG_S6</th>\n",
       "      <th>AGCAGCCGTCGAAAGC_S6</th>\n",
       "      <th>ATAGACCAGGAGTTGC_S6</th>\n",
       "      <th>GATCGTAGTTCCCGAG_S6</th>\n",
       "      <th>GGACATTTCATCTGTT_S6</th>\n",
       "      <th>ATTCTACAGGATGGTC_S6</th>\n",
       "      <th>...</th>\n",
       "      <th>TGCTACCAGTGGGTTG_S6</th>\n",
       "      <th>GAACGGAGTGCGATAG_S6</th>\n",
       "      <th>AGTGGGATCCACGCAG_S6</th>\n",
       "      <th>ACTGTCCCAGATAATG_S6</th>\n",
       "      <th>GACTACACACTCGACG_S6</th>\n",
       "      <th>GTACTCCGTCGTGGCT_S6</th>\n",
       "      <th>GGACGTCTCAGAGACG_S6</th>\n",
       "      <th>TATGCCCTCACAAACC_S6</th>\n",
       "      <th>AGGCCACCATTAGGCT_S6</th>\n",
       "      <th>TAGCCGGGTATAGGTA_S6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IL6</th>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>1.584962e+00</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STAT3</th>\n",
       "      <td>-1.146612e-06</td>\n",
       "      <td>-1.146612e-06</td>\n",
       "      <td>1.584961e+00</td>\n",
       "      <td>1.584961e+00</td>\n",
       "      <td>1.584961e+00</td>\n",
       "      <td>1.584961e+00</td>\n",
       "      <td>-1.146612e-06</td>\n",
       "      <td>-1.146612e-06</td>\n",
       "      <td>1.584961e+00</td>\n",
       "      <td>1.584961e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>2.807354e+00</td>\n",
       "      <td>-1.146612e-06</td>\n",
       "      <td>9.999989e-01</td>\n",
       "      <td>-1.146612e-06</td>\n",
       "      <td>9.999989e-01</td>\n",
       "      <td>-1.146612e-06</td>\n",
       "      <td>9.999989e-01</td>\n",
       "      <td>-1.146612e-06</td>\n",
       "      <td>-1.146612e-06</td>\n",
       "      <td>9.999989e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACE2</th>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "      <td>-7.560085e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM17</th>\n",
       "      <td>9.999991e-01</td>\n",
       "      <td>2.321927e+00</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "      <td>9.999991e-01</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "      <td>3.700439e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "      <td>1.584962e+00</td>\n",
       "      <td>9.999991e-01</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "      <td>-8.505093e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGTR1</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGF</th>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "      <td>-1.575018e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL6R</th>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>9.999996e-01</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>9.999996e-01</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>9.999996e-01</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>9.999996e-01</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>2.584962e+00</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "      <td>-3.969044e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNF</th>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>9.999990e-01</td>\n",
       "      <td>-1.042661e-06</td>\n",
       "      <td>1.584961e+00</td>\n",
       "      <td>-1.042661e-06</td>\n",
       "      <td>4.247926e+00</td>\n",
       "      <td>2.999999e+00</td>\n",
       "      <td>-1.042661e-06</td>\n",
       "      <td>-1.042661e-06</td>\n",
       "      <td>9.999990e-01</td>\n",
       "      <td>-1.042661e-06</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.042661e-06</td>\n",
       "      <td>1.999999e+00</td>\n",
       "      <td>-1.042661e-06</td>\n",
       "      <td>9.999990e-01</td>\n",
       "      <td>1.584961e+00</td>\n",
       "      <td>-1.042661e-06</td>\n",
       "      <td>2.807354e+00</td>\n",
       "      <td>9.999990e-01</td>\n",
       "      <td>-1.042661e-06</td>\n",
       "      <td>-1.042661e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NFKB1</th>\n",
       "      <td>-7.182079e-07</td>\n",
       "      <td>9.999993e-01</td>\n",
       "      <td>9.999993e-01</td>\n",
       "      <td>-7.182079e-07</td>\n",
       "      <td>9.999993e-01</td>\n",
       "      <td>2.321927e+00</td>\n",
       "      <td>-7.182079e-07</td>\n",
       "      <td>-7.182079e-07</td>\n",
       "      <td>-7.182079e-07</td>\n",
       "      <td>9.999993e-01</td>\n",
       "      <td>...</td>\n",
       "      <td>9.999993e-01</td>\n",
       "      <td>9.999993e-01</td>\n",
       "      <td>1.584962e+00</td>\n",
       "      <td>-7.182079e-07</td>\n",
       "      <td>-7.182079e-07</td>\n",
       "      <td>-7.182079e-07</td>\n",
       "      <td>9.999993e-01</td>\n",
       "      <td>-7.182079e-07</td>\n",
       "      <td>9.999993e-01</td>\n",
       "      <td>-7.182079e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCN1</th>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "      <td>-6.300071e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11 rows × 327 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ATTGGACCAGGGATTG_S6  CATCCACGTCATGCAT_S6  ACCTTTAAGAACAATC_S6  \\\n",
       "IL6           -6.300071e-09        -6.300071e-09         1.584962e+00   \n",
       "STAT3         -1.146612e-06        -1.146612e-06         1.584961e+00   \n",
       "ACE2          -7.560085e-08        -7.560085e-08        -7.560085e-08   \n",
       "ADAM17         9.999991e-01         2.321927e+00        -8.505093e-07   \n",
       "AGTR1          0.000000e+00         0.000000e+00         0.000000e+00   \n",
       "EGF           -1.575018e-08        -1.575018e-08        -1.575018e-08   \n",
       "IL6R          -3.969044e-07        -3.969044e-07         9.999996e-01   \n",
       "TNF           -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "EGFR           9.999990e-01        -1.042661e-06         1.584961e+00   \n",
       "NFKB1         -7.182079e-07         9.999993e-01         9.999993e-01   \n",
       "FCN1          -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "\n",
       "        CCTTTCTGTAGGCATG_S6  TATTACCCAAAGTGCG_S6  AGCAGCCGTCGAAAGC_S6  \\\n",
       "IL6           -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "STAT3          1.584961e+00         1.584961e+00         1.584961e+00   \n",
       "ACE2          -7.560085e-08        -7.560085e-08        -7.560085e-08   \n",
       "ADAM17        -8.505093e-07        -8.505093e-07         9.999991e-01   \n",
       "AGTR1          0.000000e+00         0.000000e+00         0.000000e+00   \n",
       "EGF           -1.575018e-08        -1.575018e-08        -1.575018e-08   \n",
       "IL6R          -3.969044e-07         9.999996e-01        -3.969044e-07   \n",
       "TNF           -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "EGFR          -1.042661e-06         4.247926e+00         2.999999e+00   \n",
       "NFKB1         -7.182079e-07         9.999993e-01         2.321927e+00   \n",
       "FCN1          -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "\n",
       "        ATAGACCAGGAGTTGC_S6  GATCGTAGTTCCCGAG_S6  GGACATTTCATCTGTT_S6  \\\n",
       "IL6           -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "STAT3         -1.146612e-06        -1.146612e-06         1.584961e+00   \n",
       "ACE2          -7.560085e-08        -7.560085e-08        -7.560085e-08   \n",
       "ADAM17        -8.505093e-07        -8.505093e-07        -8.505093e-07   \n",
       "AGTR1          0.000000e+00         0.000000e+00         0.000000e+00   \n",
       "EGF           -1.575018e-08        -1.575018e-08        -1.575018e-08   \n",
       "IL6R          -3.969044e-07         9.999996e-01        -3.969044e-07   \n",
       "TNF           -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "EGFR          -1.042661e-06        -1.042661e-06         9.999990e-01   \n",
       "NFKB1         -7.182079e-07        -7.182079e-07        -7.182079e-07   \n",
       "FCN1          -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "\n",
       "        ATTCTACAGGATGGTC_S6  ...  TGCTACCAGTGGGTTG_S6  GAACGGAGTGCGATAG_S6  \\\n",
       "IL6           -6.300071e-09  ...        -6.300071e-09        -6.300071e-09   \n",
       "STAT3          1.584961e+00  ...         2.807354e+00        -1.146612e-06   \n",
       "ACE2          -7.560085e-08  ...        -7.560085e-08        -7.560085e-08   \n",
       "ADAM17         3.700439e+00  ...        -8.505093e-07        -8.505093e-07   \n",
       "AGTR1          0.000000e+00  ...         0.000000e+00         0.000000e+00   \n",
       "EGF           -1.575018e-08  ...        -1.575018e-08        -1.575018e-08   \n",
       "IL6R          -3.969044e-07  ...        -3.969044e-07        -3.969044e-07   \n",
       "TNF           -6.300071e-09  ...        -6.300071e-09        -6.300071e-09   \n",
       "EGFR          -1.042661e-06  ...        -1.042661e-06         1.999999e+00   \n",
       "NFKB1          9.999993e-01  ...         9.999993e-01         9.999993e-01   \n",
       "FCN1          -6.300071e-09  ...        -6.300071e-09        -6.300071e-09   \n",
       "\n",
       "        AGTGGGATCCACGCAG_S6  ACTGTCCCAGATAATG_S6  GACTACACACTCGACG_S6  \\\n",
       "IL6           -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "STAT3          9.999989e-01        -1.146612e-06         9.999989e-01   \n",
       "ACE2          -7.560085e-08        -7.560085e-08        -7.560085e-08   \n",
       "ADAM17        -8.505093e-07         1.584962e+00         9.999991e-01   \n",
       "AGTR1          0.000000e+00         0.000000e+00         0.000000e+00   \n",
       "EGF           -1.575018e-08        -1.575018e-08        -1.575018e-08   \n",
       "IL6R          -3.969044e-07         9.999996e-01        -3.969044e-07   \n",
       "TNF           -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "EGFR          -1.042661e-06         9.999990e-01         1.584961e+00   \n",
       "NFKB1          1.584962e+00        -7.182079e-07        -7.182079e-07   \n",
       "FCN1          -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "\n",
       "        GTACTCCGTCGTGGCT_S6  GGACGTCTCAGAGACG_S6  TATGCCCTCACAAACC_S6  \\\n",
       "IL6           -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "STAT3         -1.146612e-06         9.999989e-01        -1.146612e-06   \n",
       "ACE2          -7.560085e-08        -7.560085e-08        -7.560085e-08   \n",
       "ADAM17        -8.505093e-07        -8.505093e-07        -8.505093e-07   \n",
       "AGTR1          0.000000e+00         0.000000e+00         0.000000e+00   \n",
       "EGF           -1.575018e-08        -1.575018e-08        -1.575018e-08   \n",
       "IL6R          -3.969044e-07         2.584962e+00        -3.969044e-07   \n",
       "TNF           -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "EGFR          -1.042661e-06         2.807354e+00         9.999990e-01   \n",
       "NFKB1         -7.182079e-07         9.999993e-01        -7.182079e-07   \n",
       "FCN1          -6.300071e-09        -6.300071e-09        -6.300071e-09   \n",
       "\n",
       "        AGGCCACCATTAGGCT_S6  TAGCCGGGTATAGGTA_S6  \n",
       "IL6           -6.300071e-09        -6.300071e-09  \n",
       "STAT3         -1.146612e-06         9.999989e-01  \n",
       "ACE2          -7.560085e-08        -7.560085e-08  \n",
       "ADAM17        -8.505093e-07        -8.505093e-07  \n",
       "AGTR1          0.000000e+00         0.000000e+00  \n",
       "EGF           -1.575018e-08        -1.575018e-08  \n",
       "IL6R          -3.969044e-07        -3.969044e-07  \n",
       "TNF           -6.300071e-09        -6.300071e-09  \n",
       "EGFR          -1.042661e-06        -1.042661e-06  \n",
       "NFKB1          9.999993e-01        -7.182079e-07  \n",
       "FCN1          -6.300071e-09        -6.300071e-09  \n",
       "\n",
       "[11 rows x 327 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "library_normalized_patient_celltype_genes['S6','Epithelial']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## All samples and cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>celltype</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">B</th>\n",
       "      <th>HC1</th>\n",
       "      <th>C51</th>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC2</th>\n",
       "      <th>C52</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC3</th>\n",
       "      <th>C100</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC4</th>\n",
       "      <th>GSM3660650</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1</th>\n",
       "      <th>C141</th>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">pDC</th>\n",
       "      <th>M3</th>\n",
       "      <th>C144</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S1</th>\n",
       "      <th>C145</th>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S2</th>\n",
       "      <th>C143</th>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S4</th>\n",
       "      <th>C148</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S6</th>\n",
       "      <th>C152</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>119 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                ID  group  disease  hasnCoV  cluster  barcode\n",
       "celltype sample_new sample                                                   \n",
       "B        HC1        C51         15     15       15       15       15       15\n",
       "         HC2        C52          5      5        5        5        5        5\n",
       "         HC3        C100         9      9        9        9        9        9\n",
       "         HC4        GSM3660650  22     22       22       22       22       22\n",
       "         M1         C141        35     35       35       35       35       35\n",
       "...                             ..    ...      ...      ...      ...      ...\n",
       "pDC      M3         C144         3      3        3        3        3        3\n",
       "         S1         C145        42     42       42       42       42       42\n",
       "         S2         C143        32     32       32       32       32       32\n",
       "         S4         C148         1      1        1        1        1        1\n",
       "         S6         C152         1      1        1        1        1        1\n",
       "\n",
       "[119 rows x 6 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_cells.groupby(['celltype','sample_new', 'sample']).count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract samples for each celltype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAACCTGAGACACTAA-1</th>\n",
       "      <th>AAACCTGAGGAGTACC-1</th>\n",
       "      <th>AAACCTGAGGATATAC-1</th>\n",
       "      <th>AAACCTGAGGTCATCT-1</th>\n",
       "      <th>AAACCTGCACGGATAG-1</th>\n",
       "      <th>AAACCTGCACGGCTAC-1</th>\n",
       "      <th>AAACCTGCAGGGAGAG-1</th>\n",
       "      <th>AAACCTGCATGCATGT-1</th>\n",
       "      <th>AAACCTGGTCTAAACC-1</th>\n",
       "      <th>AAACCTGGTGCACCAC-1</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGTCACATCCGTGG-1</th>\n",
       "      <th>TTTGTCACATGACATC-1</th>\n",
       "      <th>TTTGTCAGTATTCTCT-1</th>\n",
       "      <th>TTTGTCAGTCGAACAG-1</th>\n",
       "      <th>TTTGTCAGTCGGATCC-1</th>\n",
       "      <th>TTTGTCAGTTAAAGTG-1</th>\n",
       "      <th>TTTGTCAGTTCCACGG-1</th>\n",
       "      <th>TTTGTCATCGCTGATA-1</th>\n",
       "      <th>TTTGTCATCGGAAACG-1</th>\n",
       "      <th>TTTGTCATCTTTACGT-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 11115 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             AAACCTGAGACACTAA-1  AAACCTGAGGAGTACC-1  AAACCTGAGGATATAC-1  \\\n",
       "MIR1302-2HG                   0                   0                   0   \n",
       "FAM138A                       0                   0                   0   \n",
       "OR4F5                         0                   0                   0   \n",
       "AL627309.1                    0                   0                   0   \n",
       "AL627309.3                    0                   0                   0   \n",
       "\n",
       "             AAACCTGAGGTCATCT-1  AAACCTGCACGGATAG-1  AAACCTGCACGGCTAC-1  \\\n",
       "MIR1302-2HG                   0                   0                   0   \n",
       "FAM138A                       0                   0                   0   \n",
       "OR4F5                         0                   0                   0   \n",
       "AL627309.1                    0                   0                   0   \n",
       "AL627309.3                    0                   0                   0   \n",
       "\n",
       "             AAACCTGCAGGGAGAG-1  AAACCTGCATGCATGT-1  AAACCTGGTCTAAACC-1  \\\n",
       "MIR1302-2HG                   0                   0                   0   \n",
       "FAM138A                       0                   0                   0   \n",
       "OR4F5                         0                   0                   0   \n",
       "AL627309.1                    0                   0                   0   \n",
       "AL627309.3                    0                   0                   0   \n",
       "\n",
       "             AAACCTGGTGCACCAC-1  ...  TTTGTCACATCCGTGG-1  TTTGTCACATGACATC-1  \\\n",
       "MIR1302-2HG                   0  ...                   0                   0   \n",
       "FAM138A                       0  ...                   0                   0   \n",
       "OR4F5                         0  ...                   0                   0   \n",
       "AL627309.1                    0  ...                   0                   0   \n",
       "AL627309.3                    0  ...                   0                   0   \n",
       "\n",
       "             TTTGTCAGTATTCTCT-1  TTTGTCAGTCGAACAG-1  TTTGTCAGTCGGATCC-1  \\\n",
       "MIR1302-2HG                   0                   0                   0   \n",
       "FAM138A                       0                   0                   0   \n",
       "OR4F5                         0                   0                   0   \n",
       "AL627309.1                    0                   0                   0   \n",
       "AL627309.3                    0                   0                   0   \n",
       "\n",
       "             TTTGTCAGTTAAAGTG-1  TTTGTCAGTTCCACGG-1  TTTGTCATCGCTGATA-1  \\\n",
       "MIR1302-2HG                   0                   0                   0   \n",
       "FAM138A                       0                   0                   0   \n",
       "OR4F5                         0                   0                   0   \n",
       "AL627309.1                    0                   0                   0   \n",
       "AL627309.3                    0                   0                   0   \n",
       "\n",
       "             TTTGTCATCGGAAACG-1  TTTGTCATCTTTACGT-1  \n",
       "MIR1302-2HG                   0                   0  \n",
       "FAM138A                       0                   0  \n",
       "OR4F5                         0                   0  \n",
       "AL627309.1                    0                   0  \n",
       "AL627309.3                    0                   0  \n",
       "\n",
       "[5 rows x 11115 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw[patient].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import GEOparse\n",
    "\n",
    "def get_celltype_barcodes_of_patient( patient, celltypes, raw, all_cells, patient_to_sample, raw_barcodes ):\n",
    "    celltype_barcodes = dict()\n",
    "    celltypes_of_patient = dict()\n",
    "    for celltype in celltypes:\n",
    "        celltype_barcodes[celltype] =  get_celltype_barcodes_from_patient( patient, raw, celltype, all_cells)\n",
    "        if len(celltype_barcodes[celltype] & raw_barcodes[patient]) != len(celltype_barcodes[celltype,patient]):\n",
    "            print('{} ({}) \\traw barcodes: {}\\t{} metadata barcodes: {}\\t intersection: {}'.format(\n",
    "                patient, patient_to_sample[patient],\n",
    "                len(raw_barcodes[patient]),\n",
    "                celltype,\n",
    "                len(celltype_barcodes[celltype,patient]),\n",
    "                len(raw_barcodes[patient] & celltype_barcodes[celltype,patient])))\n",
    "    return celltype_barcodes\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "('NK', 'HC2')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-29-44c73ccd3a53>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mh2_celltype_barcodes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_celltype_barcodes_of_patient\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'HC2'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcelltypes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraw\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mall_cells\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpatient_to_sample\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraw_barcodes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mh2_celltype_barcodes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-28-6cd38ed3cd38>\u001b[0m in \u001b[0;36mget_celltype_barcodes_of_patient\u001b[0;34m(patient, celltypes, raw, all_cells, patient_to_sample, raw_barcodes)\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mcelltype\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcelltypes\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m         \u001b[0mcelltype_barcodes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcelltype\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m  \u001b[0mget_celltype_barcodes_from_patient\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0mpatient\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraw\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcelltype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mall_cells\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcelltype_barcodes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcelltype\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0mraw_barcodes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mpatient\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcelltype_barcodes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcelltype\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpatient\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m             print('{} ({}) \\traw barcodes: {}\\t{} metadata barcodes: {}\\t intersection: {}'.format(\n\u001b[1;32m     10\u001b[0m                 \u001b[0mpatient\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpatient_to_sample\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mpatient\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: ('NK', 'HC2')"
     ]
    }
   ],
   "source": [
    "h2_celltype_barcodes = get_celltype_barcodes_of_patient('HC2', celltypes, raw, all_cells, patient_to_sample, raw_barcodes)\n",
    "h2_celltype_barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import GEOparse\n",
    "geo = dict()\n",
    "raw = dict()\n",
    "raw_barcodes = dict()\n",
    "metadata_barcodes = dict()\n",
    "epithelial_barcodes = dict()\n",
    "sample_to_patient = dict(all_cells.groupby(['sample','sample_new']).count().index.tolist())\n",
    "for gsm_file in balf_files:\n",
    "    gsm_id, sample = gsm_file.split('_')[0:2]\n",
    "    patient = sample_to_patient[sample]\n",
    "    geo[patient] = GEOparse.get_GEO(geo=gsm_id, destdir=balf_dir)\n",
    "    raw[patient] = get_matrix_from_h5(os.path.join(balf_dir,gsm_file))\n",
    "    metadata_barcodes[patient] = set(all_cells[all_cells['sample_new'] == patient]['ID']\n",
    "                                    .str.split('_').str.get(0)\n",
    "                                    .values.tolist())\n",
    "    raw_barcodes[patient] = set([raw_barcode.split('-')[0] for raw_barcode in raw[patient].columns])\n",
    "    print('{} ({})\\traw barcodes: {}\\tepithelial metadata barcodes: {}\\t intersection: {}'.format(\n",
    "                 patient, sample,\n",
    "                len(raw_barcodes[patient]), \n",
    "                len(metadata_barcodes[patient]), \n",
    "                len(raw_barcodes[patient] & metadata_barcodes[patient])))\n",
    "    epithelial_barcodes[patient] = set(all_cells[(all_cells['sample_new'] == patient) &\n",
    "                                               (all_cells['celltype'] == 'Epithelial')]['ID']\n",
    "                                    .str.split('_').str.get(0)\n",
    "                                    .values.tolist())\n",
    "    print('{} ({}) \\traw barcodes: {}\\tepithelial metadata barcodes: {}\\t intersection: {}'.format(\n",
    "                patient, sample,\n",
    "                len(raw_barcodes[patient]),\n",
    "                len(epithelial_barcodes[patient]),\n",
    "                len(raw_barcodes[patient] & epithelial_barcodes[patient])))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Healthy control 4 (HC4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>celltype</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19173</th>\n",
       "      <td>AAACCTGCAGCATGAG_4</td>\n",
       "      <td>GSM3660650</td>\n",
       "      <td>HC4</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>6</td>\n",
       "      <td>T</td>\n",
       "      <td>AAACCTGCAGCATGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19174</th>\n",
       "      <td>AAACCTGGTATATGGA_4</td>\n",
       "      <td>GSM3660650</td>\n",
       "      <td>HC4</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>13</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>AAACCTGGTATATGGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19175</th>\n",
       "      <td>AAACCTGGTTGTACAC_4</td>\n",
       "      <td>GSM3660650</td>\n",
       "      <td>HC4</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>25</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>AAACCTGGTTGTACAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19176</th>\n",
       "      <td>AAACCTGTCAGAGGTG_4</td>\n",
       "      <td>GSM3660650</td>\n",
       "      <td>HC4</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>AAACCTGTCAGAGGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19177</th>\n",
       "      <td>AAACCTGTCGACAGCC_4</td>\n",
       "      <td>GSM3660650</td>\n",
       "      <td>HC4</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>AAACCTGTCGACAGCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       ID      sample sample_new group disease hasnCoV  \\\n",
       "19173  AAACCTGCAGCATGAG_4  GSM3660650        HC4    HC       N       N   \n",
       "19174  AAACCTGGTATATGGA_4  GSM3660650        HC4    HC       N       N   \n",
       "19175  AAACCTGGTTGTACAC_4  GSM3660650        HC4    HC       N       N   \n",
       "19176  AAACCTGTCAGAGGTG_4  GSM3660650        HC4    HC       N       N   \n",
       "19177  AAACCTGTCGACAGCC_4  GSM3660650        HC4    HC       N       N   \n",
       "\n",
       "       cluster    celltype           barcode  \n",
       "19173        6           T  AAACCTGCAGCATGAG  \n",
       "19174       13  Epithelial  AAACCTGGTATATGGA  \n",
       "19175       25  Epithelial  AAACCTGGTTGTACAC  \n",
       "19176       16  Epithelial  AAACCTGTCAGAGGTG  \n",
       "19177       16  Epithelial  AAACCTGTCGACAGCC  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hc4_cells = all_cells[(all_cells['sample_new'] == 'HC4')]\n",
    "hc4_cells.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>celltype</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epithelial</th>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Macrophages</th>\n",
       "      <td>683</td>\n",
       "      <td>683</td>\n",
       "      <td>683</td>\n",
       "      <td>683</td>\n",
       "      <td>683</td>\n",
       "      <td>683</td>\n",
       "      <td>683</td>\n",
       "      <td>683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mast</th>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK</th>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>533</td>\n",
       "      <td>533</td>\n",
       "      <td>533</td>\n",
       "      <td>533</td>\n",
       "      <td>533</td>\n",
       "      <td>533</td>\n",
       "      <td>533</td>\n",
       "      <td>533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mDC</th>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pDC</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               ID  sample  sample_new  group  disease  hasnCoV  cluster  \\\n",
       "celltype                                                                  \n",
       "B              22      22          22     22       22       22       22   \n",
       "Epithelial   1383    1383        1383   1383     1383     1383     1383   \n",
       "Macrophages   683     683         683    683      683      683      683   \n",
       "Mast           19      19          19     19       19       19       19   \n",
       "NK             29      29          29     29       29       29       29   \n",
       "T             533     533         533    533      533      533      533   \n",
       "mDC            37      37          37     37       37       37       37   \n",
       "pDC             4       4           4      4        4        4        4   \n",
       "\n",
       "             barcode  \n",
       "celltype              \n",
       "B                 22  \n",
       "Epithelial      1383  \n",
       "Macrophages      683  \n",
       "Mast              19  \n",
       "NK                29  \n",
       "T                533  \n",
       "mDC               37  \n",
       "pDC                4  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hc4_cells.groupby('celltype').count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse Healthy control 4 (HC4) GEO Sample  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "05-Jun-2020 15:27:49 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "05-Jun-2020 15:27:49 INFO GEOparse - File already exist: using local version.\n",
      "05-Jun-2020 15:27:49 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM3660650.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " - title : SC249NORbal_fresh\n",
      " - geo_accession : GSM3660650\n",
      " - status : Public on Jun 21 2019\n",
      " - submission_date : Mar 07 2019\n",
      " - last_update_date : Jun 23 2019\n",
      " - type : SRA\n",
      " - channel_count : 1\n",
      " - source_name_ch1 : fresh Brochoalvelolar lavage (BAL) + brush cells from healthy donor\n",
      " - organism_ch1 : Homo sapiens\n",
      " - taxid_ch1 : 9606\n",
      " - characteristics_ch1 : subject status: normal/healthy donor, tissue: lung, lung section/portion: fresh Brochoalvelolar lavage (BAL) + brush cells\n",
      " - molecule_ch1 : polyA RNA\n",
      " - extract_protocol_ch1 : 10X Genomics 3 Prime V2 chemistry, Explanted lungs were finely diced and digested for 1 hour in a mix of collagenase A/DNAse I.  A single cell suspension was generated by filtering lungs through 70 micron filters and resuspending cells in PBS containing 0.04% BSA.  Cells were loaded into the 10X Genomics Chromium instrument per manufacturer's directions., Libraries were constructed according to 10X Genomics' Single Cell 3’ Reagent Kits v2 User Guide\n",
      " - data_processing : 10X Genomics' Cell Ranger 2.0 software's mkfastq command was used for basecalling.  It is a wrapper for Illumina's bcl2fastq, Sequenced reads were then mapped to GRCh38 whole genome using 10X Genomics' Cell Ranger 2.0 software's cellranger count function, Cell Ranger output of raw_gene_bc_matrices were utilized for further downstream analysis in R with the package Seurat 2.3.4, only cells expressing at least 200 genes were kept for downstream analysis.  Further filtration removed cells expressing greater than 50,000 UMI or 35% mitochondrial genes., Genome_build: GRCh38, Supplementary_files_format_and_content: tsv files contain lists of genes and barcodes for each sample, mtx files count gene UMI couns for each sample\n",
      " - platform_id : GPL18573\n",
      " - contact_name : Tracy,,Tabib\n",
      " - contact_laboratory : Laboratory of Robert Lafyatis\n",
      " - contact_department : Department of Medicine\n",
      " - contact_institute : University of Pittsburgh\n",
      " - contact_address : 200 Lothrop Street\n",
      " - contact_city : Pittsburgh\n",
      " - contact_state : PA\n",
      " - contact_zip/postal_code : 15213\n",
      " - contact_country : USA\n",
      " - instrument_model : Illumina NextSeq 500\n",
      " - library_selection : cDNA\n",
      " - library_source : transcriptomic\n",
      " - library_strategy : RNA-Seq\n",
      " - relation : BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN11086899\n",
      " - supplementary_file_1 : ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3660nnn/GSM3660650/suppl/GSM3660650_SC249NORbal_fresh_barcodes.tsv.gz\n",
      " - supplementary_file_2 : ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3660nnn/GSM3660650/suppl/GSM3660650_SC249NORbal_fresh_genes.tsv.gz\n",
      " - supplementary_file_3 : ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3660nnn/GSM3660650/suppl/GSM3660650_SC249NORbal_fresh_matrix.mtx.gz\n",
      " - series_id : GSE128033\n",
      " - data_row_count : 0\n",
      "Table data:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "import GEOparse\n",
    "\n",
    "HC4_geo_id = \"GSM3660650\"\n",
    "HC4_geo_sample = GEOparse.get_GEO(geo=HC4_geo_id, destdir=balf_dir)\n",
    "\n",
    "for key, value in HC4_geo_sample.metadata.items():\n",
    "        print(\" - %s : %s\" % (key, \", \".join(value)))\n",
    "print (\"Table data:\",)\n",
    "print (HC4_geo_sample.table.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>barcode</th>\n",
       "      <th>AAACCTGAGAAACCAT-1</th>\n",
       "      <th>AAACCTGAGAAACCGC-1</th>\n",
       "      <th>AAACCTGAGAAACCTA-1</th>\n",
       "      <th>AAACCTGAGAAACGAG-1</th>\n",
       "      <th>AAACCTGAGAAACGCC-1</th>\n",
       "      <th>AAACCTGAGAAAGTGG-1</th>\n",
       "      <th>AAACCTGAGAACAACT-1</th>\n",
       "      <th>AAACCTGAGAACAATC-1</th>\n",
       "      <th>AAACCTGAGAACTCGG-1</th>\n",
       "      <th>AAACCTGAGAACTGTA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL732372.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F29</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC114498.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "barcode      AAACCTGAGAAACCAT-1  AAACCTGAGAAACCGC-1  AAACCTGAGAAACCTA-1  \\\n",
       "gene_name                                                                 \n",
       "MIR1302-2HG                   0                   0                   0   \n",
       "FAM138A                       0                   0                   0   \n",
       "OR4F5                         0                   0                   0   \n",
       "AL627309.1                    0                   0                   0   \n",
       "AL627309.3                    0                   0                   0   \n",
       "AL627309.2                    0                   0                   0   \n",
       "AL627309.4                    0                   0                   0   \n",
       "AL732372.1                    0                   0                   0   \n",
       "OR4F29                        0                   0                   0   \n",
       "AC114498.1                    0                   0                   0   \n",
       "\n",
       "barcode      AAACCTGAGAAACGAG-1  AAACCTGAGAAACGCC-1  AAACCTGAGAAAGTGG-1  \\\n",
       "gene_name                                                                 \n",
       "MIR1302-2HG                   0                   0                   0   \n",
       "FAM138A                       0                   0                   0   \n",
       "OR4F5                         0                   0                   0   \n",
       "AL627309.1                    0                   0                   0   \n",
       "AL627309.3                    0                   0                   0   \n",
       "AL627309.2                    0                   0                   0   \n",
       "AL627309.4                    0                   0                   0   \n",
       "AL732372.1                    0                   0                   0   \n",
       "OR4F29                        0                   0                   0   \n",
       "AC114498.1                    0                   0                   0   \n",
       "\n",
       "barcode      AAACCTGAGAACAACT-1  AAACCTGAGAACAATC-1  AAACCTGAGAACTCGG-1  \\\n",
       "gene_name                                                                 \n",
       "MIR1302-2HG                   0                   0                   0   \n",
       "FAM138A                       0                   0                   0   \n",
       "OR4F5                         0                   0                   0   \n",
       "AL627309.1                    0                   0                   0   \n",
       "AL627309.3                    0                   0                   0   \n",
       "AL627309.2                    0                   0                   0   \n",
       "AL627309.4                    0                   0                   0   \n",
       "AL732372.1                    0                   0                   0   \n",
       "OR4F29                        0                   0                   0   \n",
       "AC114498.1                    0                   0                   0   \n",
       "\n",
       "barcode      AAACCTGAGAACTGTA-1  \n",
       "gene_name                        \n",
       "MIR1302-2HG                   0  \n",
       "FAM138A                       0  \n",
       "OR4F5                         0  \n",
       "AL627309.1                    0  \n",
       "AL627309.3                    0  \n",
       "AL627309.2                    0  \n",
       "AL627309.4                    0  \n",
       "AL732372.1                    0  \n",
       "OR4F29                        0  \n",
       "AC114498.1                    0  "
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#print(hc4_raw.iloc[:10,:10].to_html())\n",
    "hc4_raw.iloc[:10,:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Macrophage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>celltype</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGACACTAA_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGAGACACTAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGGAGTACC_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGAGGAGTACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGAGGATATAC_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGAGGATATAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGAGGTCATCT_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGAGGTCATCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGCACGGATAG_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>5</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGCACGGATAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   ID sample sample_new group disease hasnCoV  cluster  \\\n",
       "0  AAACCTGAGACACTAA_1    C51        HC1    HC       N       N        3   \n",
       "1  AAACCTGAGGAGTACC_1    C51        HC1    HC       N       N        3   \n",
       "2  AAACCTGAGGATATAC_1    C51        HC1    HC       N       N        3   \n",
       "3  AAACCTGAGGTCATCT_1    C51        HC1    HC       N       N        3   \n",
       "4  AAACCTGCACGGATAG_1    C51        HC1    HC       N       N        5   \n",
       "\n",
       "      celltype           barcode  \n",
       "0  Macrophages  AAACCTGAGACACTAA  \n",
       "1  Macrophages  AAACCTGAGGAGTACC  \n",
       "2  Macrophages  AAACCTGAGGATATAC  \n",
       "3  Macrophages  AAACCTGAGGTCATCT  \n",
       "4  Macrophages  AAACCTGCACGGATAG  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "macrophage = all_cells[all_cells['celltype']== 'Macrophages']\n",
    "macrophage.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2710"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Epithelial cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>celltype</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1070</th>\n",
       "      <td>AGATTGCAGATAGCAT_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>AGATTGCAGATAGCAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1313</th>\n",
       "      <td>AGGGAGTGTTGATTGC_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>AGGGAGTGTTGATTGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1445</th>\n",
       "      <td>AGTGGGACACCAGATT_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>AGTGGGACACCAGATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1799</th>\n",
       "      <td>ATTACTCTCGGAAATA_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>ATTACTCTCGGAAATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2729</th>\n",
       "      <td>CCATGTCCATCCGCGA_1</td>\n",
       "      <td>C51</td>\n",
       "      <td>HC1</td>\n",
       "      <td>HC</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>CCATGTCCATCCGCGA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ID sample sample_new group disease hasnCoV  cluster  \\\n",
       "1070  AGATTGCAGATAGCAT_1    C51        HC1    HC       N       N       16   \n",
       "1313  AGGGAGTGTTGATTGC_1    C51        HC1    HC       N       N       16   \n",
       "1445  AGTGGGACACCAGATT_1    C51        HC1    HC       N       N       16   \n",
       "1799  ATTACTCTCGGAAATA_1    C51        HC1    HC       N       N       16   \n",
       "2729  CCATGTCCATCCGCGA_1    C51        HC1    HC       N       N       16   \n",
       "\n",
       "        celltype           barcode  \n",
       "1070  Epithelial  AGATTGCAGATAGCAT  \n",
       "1313  Epithelial  AGGGAGTGTTGATTGC  \n",
       "1445  Epithelial  AGTGGGACACCAGATT  \n",
       "1799  Epithelial  ATTACTCTCGGAAATA  \n",
       "2729  Epithelial  CCATGTCCATCCGCGA  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epithelial = all_cells[all_cells['celltype'] == 'Epithelial']\n",
    "epithelial.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>celltype</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21908</th>\n",
       "      <td>AAAGCAAAGTGAAGAG_5</td>\n",
       "      <td>C141</td>\n",
       "      <td>M1</td>\n",
       "      <td>M</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>AAAGCAAAGTGAAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22026</th>\n",
       "      <td>AAGGAGCTCTAAGCCA_5</td>\n",
       "      <td>C141</td>\n",
       "      <td>M1</td>\n",
       "      <td>M</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>AAGGAGCTCTAAGCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22071</th>\n",
       "      <td>ACACCCTAGTTACCCA_5</td>\n",
       "      <td>C141</td>\n",
       "      <td>M1</td>\n",
       "      <td>M</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>13</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>ACACCCTAGTTACCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22129</th>\n",
       "      <td>ACCAGTAAGTTTCCTT_5</td>\n",
       "      <td>C141</td>\n",
       "      <td>M1</td>\n",
       "      <td>M</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>ACCAGTAAGTTTCCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22173</th>\n",
       "      <td>ACGAGGATCAACCAAC_5</td>\n",
       "      <td>C141</td>\n",
       "      <td>M1</td>\n",
       "      <td>M</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>ACGAGGATCAACCAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25138</th>\n",
       "      <td>TGCTACCCAACCGCCA_5</td>\n",
       "      <td>C141</td>\n",
       "      <td>M1</td>\n",
       "      <td>M</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>TGCTACCCAACCGCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25152</th>\n",
       "      <td>TGGACGCAGGGATACC_5</td>\n",
       "      <td>C141</td>\n",
       "      <td>M1</td>\n",
       "      <td>M</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>TGGACGCAGGGATACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25222</th>\n",
       "      <td>TGTCCCATCCGCGCAA_5</td>\n",
       "      <td>C141</td>\n",
       "      <td>M1</td>\n",
       "      <td>M</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>16</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>TGTCCCATCCGCGCAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25330</th>\n",
       "      <td>TTCTCCTTCCCTAACC_5</td>\n",
       "      <td>C141</td>\n",
       "      <td>M1</td>\n",
       "      <td>M</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>25</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>TTCTCCTTCCCTAACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25383</th>\n",
       "      <td>TTTACTGAGATGGGTC_5</td>\n",
       "      <td>C141</td>\n",
       "      <td>M1</td>\n",
       "      <td>M</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>13</td>\n",
       "      <td>Epithelial</td>\n",
       "      <td>TTTACTGAGATGGGTC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>83 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       ID sample sample_new group disease hasnCoV  cluster  \\\n",
       "21908  AAAGCAAAGTGAAGAG_5   C141         M1     M       Y       N       16   \n",
       "22026  AAGGAGCTCTAAGCCA_5   C141         M1     M       Y       N       16   \n",
       "22071  ACACCCTAGTTACCCA_5   C141         M1     M       Y       N       13   \n",
       "22129  ACCAGTAAGTTTCCTT_5   C141         M1     M       Y       N       16   \n",
       "22173  ACGAGGATCAACCAAC_5   C141         M1     M       Y       N       16   \n",
       "...                   ...    ...        ...   ...     ...     ...      ...   \n",
       "25138  TGCTACCCAACCGCCA_5   C141         M1     M       Y       N       16   \n",
       "25152  TGGACGCAGGGATACC_5   C141         M1     M       Y       N       16   \n",
       "25222  TGTCCCATCCGCGCAA_5   C141         M1     M       Y       N       16   \n",
       "25330  TTCTCCTTCCCTAACC_5   C141         M1     M       Y       N       25   \n",
       "25383  TTTACTGAGATGGGTC_5   C141         M1     M       Y       N       13   \n",
       "\n",
       "         celltype           barcode  \n",
       "21908  Epithelial  AAAGCAAAGTGAAGAG  \n",
       "22026  Epithelial  AAGGAGCTCTAAGCCA  \n",
       "22071  Epithelial  ACACCCTAGTTACCCA  \n",
       "22129  Epithelial  ACCAGTAAGTTTCCTT  \n",
       "22173  Epithelial  ACGAGGATCAACCAAC  \n",
       "...           ...               ...  \n",
       "25138  Epithelial  TGCTACCCAACCGCCA  \n",
       "25152  Epithelial  TGGACGCAGGGATACC  \n",
       "25222  Epithelial  TGTCCCATCCGCGCAA  \n",
       "25330  Epithelial  TTCTCCTTCCCTAACC  \n",
       "25383  Epithelial  TTTACTGAGATGGGTC  \n",
       "\n",
       "[83 rows x 9 columns]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epithelial[epithelial['sample'] == 'C141']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Severe COVID-19 patient 2 (S2) Epithelial cell barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "309"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "S2_epi = set(epithelial.loc[epithelial['sample_new'] == 'S2']['barcode'].values.tolist())\n",
    "len(S2_epi)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Healthy control 4 (HC4) Epithelial cell barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1383"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HC4_epi = set(epithelial.loc[epithelial['sample_new']=='HC4']['barcode'].values.tolist())\n",
    "len(HC4_epi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>celltype</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C100</th>\n",
       "      <th>HC3</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C141</th>\n",
       "      <th>M1</th>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C142</th>\n",
       "      <th>M2</th>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C143</th>\n",
       "      <th>S2</th>\n",
       "      <td>309</td>\n",
       "      <td>309</td>\n",
       "      <td>309</td>\n",
       "      <td>309</td>\n",
       "      <td>309</td>\n",
       "      <td>309</td>\n",
       "      <td>309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C144</th>\n",
       "      <th>M3</th>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C145</th>\n",
       "      <th>S1</th>\n",
       "      <td>368</td>\n",
       "      <td>368</td>\n",
       "      <td>368</td>\n",
       "      <td>368</td>\n",
       "      <td>368</td>\n",
       "      <td>368</td>\n",
       "      <td>368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C146</th>\n",
       "      <th>S3</th>\n",
       "      <td>312</td>\n",
       "      <td>312</td>\n",
       "      <td>312</td>\n",
       "      <td>312</td>\n",
       "      <td>312</td>\n",
       "      <td>312</td>\n",
       "      <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C148</th>\n",
       "      <th>S4</th>\n",
       "      <td>373</td>\n",
       "      <td>373</td>\n",
       "      <td>373</td>\n",
       "      <td>373</td>\n",
       "      <td>373</td>\n",
       "      <td>373</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C149</th>\n",
       "      <th>S5</th>\n",
       "      <td>185</td>\n",
       "      <td>185</td>\n",
       "      <td>185</td>\n",
       "      <td>185</td>\n",
       "      <td>185</td>\n",
       "      <td>185</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C152</th>\n",
       "      <th>S6</th>\n",
       "      <td>327</td>\n",
       "      <td>327</td>\n",
       "      <td>327</td>\n",
       "      <td>327</td>\n",
       "      <td>327</td>\n",
       "      <td>327</td>\n",
       "      <td>327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C51</th>\n",
       "      <th>HC1</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C52</th>\n",
       "      <th>HC2</th>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3660650</th>\n",
       "      <th>HC4</th>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         ID  group  disease  hasnCoV  cluster  celltype  \\\n",
       "sample     sample_new                                                     \n",
       "C100       HC3           13     13       13       13       13        13   \n",
       "C141       M1            83     83       83       83       83        83   \n",
       "C142       M2            94     94       94       94       94        94   \n",
       "C143       S2           309    309      309      309      309       309   \n",
       "C144       M3            38     38       38       38       38        38   \n",
       "C145       S1           368    368      368      368      368       368   \n",
       "C146       S3           312    312      312      312      312       312   \n",
       "C148       S4           373    373      373      373      373       373   \n",
       "C149       S5           185    185      185      185      185       185   \n",
       "C152       S6           327    327      327      327      327       327   \n",
       "C51        HC1           13     13       13       13       13        13   \n",
       "C52        HC2           33     33       33       33       33        33   \n",
       "GSM3660650 HC4         1383   1383     1383     1383     1383      1383   \n",
       "\n",
       "                       barcode  \n",
       "sample     sample_new           \n",
       "C100       HC3              13  \n",
       "C141       M1               83  \n",
       "C142       M2               94  \n",
       "C143       S2              309  \n",
       "C144       M3               38  \n",
       "C145       S1              368  \n",
       "C146       S3              312  \n",
       "C148       S4              373  \n",
       "C149       S5              185  \n",
       "C152       S6              327  \n",
       "C51        HC1              13  \n",
       "C52        HC2              33  \n",
       "GSM3660650 HC4            1383  "
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epithelial.groupby(['sample','sample_new']).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('C100', 'HC3'),\n",
       " ('C141', 'M1'),\n",
       " ('C142', 'M2'),\n",
       " ('C143', 'S2'),\n",
       " ('C144', 'M3'),\n",
       " ('C145', 'S1'),\n",
       " ('C146', 'S3'),\n",
       " ('C148', 'S4'),\n",
       " ('C149', 'S5'),\n",
       " ('C152', 'S6'),\n",
       " ('C51', 'HC1'),\n",
       " ('C52', 'HC2'),\n",
       " ('GSM3660650', 'HC4')]"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epithelial.groupby(['sample','sample_new']).count().index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>celltype</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HC</th>\n",
       "      <td>1442</td>\n",
       "      <td>1442</td>\n",
       "      <td>1442</td>\n",
       "      <td>1442</td>\n",
       "      <td>1442</td>\n",
       "      <td>1442</td>\n",
       "      <td>1442</td>\n",
       "      <td>1442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>215</td>\n",
       "      <td>215</td>\n",
       "      <td>215</td>\n",
       "      <td>215</td>\n",
       "      <td>215</td>\n",
       "      <td>215</td>\n",
       "      <td>215</td>\n",
       "      <td>215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>1874</td>\n",
       "      <td>1874</td>\n",
       "      <td>1874</td>\n",
       "      <td>1874</td>\n",
       "      <td>1874</td>\n",
       "      <td>1874</td>\n",
       "      <td>1874</td>\n",
       "      <td>1874</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID  sample  sample_new  disease  hasnCoV  cluster  celltype  barcode\n",
       "group                                                                        \n",
       "HC     1442    1442        1442     1442     1442     1442      1442     1442\n",
       "M       215     215         215      215      215      215       215      215\n",
       "S      1874    1874        1874     1874     1874     1874      1874     1874"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epithelial.groupby('group').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_new</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HC1</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC2</th>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC3</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC4</th>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1</th>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M2</th>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M3</th>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S1</th>\n",
       "      <td>368</td>\n",
       "      <td>368</td>\n",
       "      <td>368</td>\n",
       "      <td>368</td>\n",
       "      <td>368</td>\n",
       "      <td>368</td>\n",
       "      <td>368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S2</th>\n",
       "      <td>309</td>\n",
       "      <td>309</td>\n",
       "      <td>309</td>\n",
       "      <td>309</td>\n",
       "      <td>309</td>\n",
       "      <td>309</td>\n",
       "      <td>309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S3</th>\n",
       "      <td>312</td>\n",
       "      <td>312</td>\n",
       "      <td>312</td>\n",
       "      <td>312</td>\n",
       "      <td>312</td>\n",
       "      <td>312</td>\n",
       "      <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S4</th>\n",
       "      <td>373</td>\n",
       "      <td>373</td>\n",
       "      <td>373</td>\n",
       "      <td>373</td>\n",
       "      <td>373</td>\n",
       "      <td>373</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S5</th>\n",
       "      <td>185</td>\n",
       "      <td>185</td>\n",
       "      <td>185</td>\n",
       "      <td>185</td>\n",
       "      <td>185</td>\n",
       "      <td>185</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S6</th>\n",
       "      <td>327</td>\n",
       "      <td>327</td>\n",
       "      <td>327</td>\n",
       "      <td>327</td>\n",
       "      <td>327</td>\n",
       "      <td>327</td>\n",
       "      <td>327</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID  sample  group  disease  hasnCoV  cluster  celltype\n",
       "sample_new                                                          \n",
       "HC1           13      13     13       13       13       13        13\n",
       "HC2           33      33     33       33       33       33        33\n",
       "HC3           13      13     13       13       13       13        13\n",
       "HC4         1383    1383   1383     1383     1383     1383      1383\n",
       "M1            83      83     83       83       83       83        83\n",
       "M2            94      94     94       94       94       94        94\n",
       "M3            38      38     38       38       38       38        38\n",
       "S1           368     368    368      368      368      368       368\n",
       "S2           309     309    309      309      309      309       309\n",
       "S3           312     312    312      312      312      312       312\n",
       "S4           373     373    373      373      373      373       373\n",
       "S5           185     185    185      185      185      185       185\n",
       "S6           327     327    327      327      327      327       327"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epithelial.groupby('sample_new').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'13': 'Epithelial',\n",
       " '16': 'Epithelial',\n",
       " '25': 'Epithelial',\n",
       " '28': 'Epithelial',\n",
       " '31': 'Epithelial'}"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epi_cluster_groups =  {'13' : 'Epithelial',\n",
    "           '16' : 'Epithelial',\n",
    "           '25' : 'Epithelial',\n",
    "           '28' : 'Epithelial',\n",
    "           '31' : 'Epithelial'}\n",
    "epi_cluster_groups"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare Raw HC4 barcodes with Metdata HC4 barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(737280, 2710, 2710)"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hc4_metadata_barcodes = set(hc4_cells['ID'].str.split('_').str.get(0).values.tolist())\n",
    "hc4_raw_barcodes = set([raw.split('-')[0] for raw in hc4_raw.columns])\n",
    "len(hc4_raw_barcodes), len(hc4_metadata_barcodes), len(hc4_raw_barcodes & hc4_metadata_barcodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Number of Epithelial cells in H4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1383, 1383)"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hc4_epithelial_cells = hc4_cells[hc4_cells['celltype'] == 'Epithelial']\n",
    "hc4_epithelial_metadata_barcodes = set(hc4_epithelial_cells['ID'].str.split('_').str.get(0).values.tolist())\n",
    "len(hc4_epithelial_metadata_barcodes), len(hc4_raw_barcodes & hc4_epithelial_metadata_barcodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write out Healthy Control 4 epithelial cells to csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_hc4_raw_columns = [b for b in hc4_raw.columns]\n",
    "hc4_raw.columns = [barcode.split('-')[0] for barcode in hc4_raw.columns]\n",
    "hc4_raw[hc4_epithelial_metadata_barcodes].to_csv(\n",
    "    os.path.join(balf_dir, 'Epithelial','hc4_umi_epithelial.csv'))\n",
    "hc4_raw.columns = old_hc4_raw_columns "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Severe COVID-19 patient 2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>celltype</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29194</th>\n",
       "      <td>AAACCTGAGACTTTCG_8</td>\n",
       "      <td>C143</td>\n",
       "      <td>S2</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>4</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGAGACTTTCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29195</th>\n",
       "      <td>AAACCTGAGCATGGCA_8</td>\n",
       "      <td>C143</td>\n",
       "      <td>S2</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>0</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGAGCATGGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29196</th>\n",
       "      <td>AAACCTGAGCCATCGC_8</td>\n",
       "      <td>C143</td>\n",
       "      <td>S2</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>3</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGAGCCATCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29197</th>\n",
       "      <td>AAACCTGAGCGTGAAC_8</td>\n",
       "      <td>C143</td>\n",
       "      <td>S2</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>15</td>\n",
       "      <td>Neutrophil</td>\n",
       "      <td>AAACCTGAGCGTGAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29198</th>\n",
       "      <td>AAACCTGAGGACATTA_8</td>\n",
       "      <td>C143</td>\n",
       "      <td>S2</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>5</td>\n",
       "      <td>Macrophages</td>\n",
       "      <td>AAACCTGAGGACATTA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       ID sample sample_new group disease hasnCoV  cluster  \\\n",
       "29194  AAACCTGAGACTTTCG_8   C143         S2     S       Y       N        4   \n",
       "29195  AAACCTGAGCATGGCA_8   C143         S2     S       Y       N        0   \n",
       "29196  AAACCTGAGCCATCGC_8   C143         S2     S       Y       N        3   \n",
       "29197  AAACCTGAGCGTGAAC_8   C143         S2     S       Y       N       15   \n",
       "29198  AAACCTGAGGACATTA_8   C143         S2     S       Y       N        5   \n",
       "\n",
       "          celltype           barcode  \n",
       "29194  Macrophages  AAACCTGAGACTTTCG  \n",
       "29195  Macrophages  AAACCTGAGCATGGCA  \n",
       "29196  Macrophages  AAACCTGAGCCATCGC  \n",
       "29197   Neutrophil  AAACCTGAGCGTGAAC  \n",
       "29198  Macrophages  AAACCTGAGGACATTA  "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s2_cells = all_cells[(all_cells['sample_new'] == 'S2')]\n",
    "\n",
    "s2_cells.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse Severe COVID-19 patient 2 (S2) GEO sample file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "S2_sample_id = 'C144'\n",
    "S2_geo_id = 'GSM4339772'\n",
    "S2_raw_file = balf_files[3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "28-May-2020 10:48:47 DEBUG utils - Directory ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/ already exists. Skipping.\n",
      "28-May-2020 10:48:47 INFO GEOparse - Downloading http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?targ=self&acc=GSM4339772&form=text&view=full to ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4339772.txt\n",
      "28-May-2020 10:48:48 DEBUG downloader - Total size: 0\n",
      "28-May-2020 10:48:48 DEBUG downloader - md5: None\n",
      "3.92kB [00:00, 1.58MB/s]\n",
      "28-May-2020 10:48:48 DEBUG downloader - Moving /var/folders/z1/nhct2gps3y94tkbwqvvd7y_jw6p1l8/T/tmpfvirto9h to /Users/zuck016/Projects/Proposals/MARS/Pathogenesis/MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4339772.txt\n",
      "28-May-2020 10:48:48 DEBUG downloader - Successfully downloaded http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?targ=self&acc=GSM4339772&form=text&view=full\n",
      "28-May-2020 10:48:48 INFO GEOparse - Parsing ../../MARS-Pathogenesis/Coronavirus/SARS-CoV-2/COVID-BALF/GSM4339772.txt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " - title : BALF, C144 (scRNA-seq)\n",
      " - geo_accession : GSM4339772\n",
      " - status : Public on Apr 22 2020\n",
      " - submission_date : Feb 25 2020\n",
      " - last_update_date : Apr 22 2020\n",
      " - type : SRA\n",
      " - channel_count : 1\n",
      " - source_name_ch1 : BALF\n",
      " - organism_ch1 : Homo sapiens\n",
      " - taxid_ch1 : 9606\n",
      " - characteristics_ch1 : disease state: Coronavirus infected disease-19 (COVID-19), patient group: mild, tissue: lung, sample type: BALF, cell subsets: Total cell\n",
      " - molecule_ch1 : polyA RNA\n",
      " - extract_protocol_ch1 : Total 11 µl of single cell suspension and 40 µl barcoded Gel Beads were loaded to Chromium Chip A to generate single-cell gel bead-in-emulsion (GEM). The poly-adenylated transcripts were reverse-transcribed later. The single-cell capturing and downstream library constructions were performed using the Chromium Single Cell 5’ library preparation kit according to the manufacturer’s protocol (10x Genomics). Full-length cDNA along with cell-barcode identifiers were PCR-amplified and sequencing libraries were prepared and normalized to 3 nM.\n",
      " - description : Single-cell RNAseq data\n",
      " - data_processing : The Cell Ranger Software Suite (Version 3.1.0) was used to perform sample de-multiplexing, barcode processing and single-cell 5’ UMI counting with human GRCh38 as the reference genome. Specifically, splicing-aware aligner STAR was used in FASTQs alignment. Cell barcodes were then determined based on distribution of UMI count automatically. Finally, gene-barcode matrix of all 6 donors and 8 previously reported healthy control was integrated with Seurat v3 to remove batch effect across different donors. Following criteria were then applied to each cell, i.e., gene number between 200 and 6000, UMI count above 1000 and mitochondrial gene percentage below 0.1., The filtered gene-barcode matrix was normalized with LogNormalize methods in Seurat and analyzed by principal component analysis (PCA) using the top 2, 000 most variable genes. Then Uniform Manifold Approximation and Projection (UMAP) was performed on the top 50 principal components for visualizing the cells. Meanwhile, graph-based clustering was performed on the PCA-reduced data for clustering analysis with Seurat v3., MAST in Seurat v3 was used to perform differential analysis. For each cluster, differentially-expressed genes (DEGs) were generated relative to all of the other cells., The TCR sequences for each single T cell were assembled by Cell Ranger vdj pipeline (v3.1.0), leading to the identification of CDR3 sequence and the rearranged TCR gene. Cells with both TCR alpha and beta chains were kept and cells with only one TCR chain were discarded., Genome_build: GRCh38 + COVID(MN908947), Supplementary_files_format_and_content: h5, raw count matrix generated by Cell Ranger count; csv, tcr contig annotations generated by Cell Ranger vdj\n",
      " - platform_id : GPL23227\n",
      " - contact_name : Zheng,,Zhang\n",
      " - contact_institute : Shenzhen 3rd People's Hospital\n",
      " - contact_address : No. 29, Bulan Road\n",
      " - contact_city : Shenzhen\n",
      " - contact_state : Guangdong\n",
      " - contact_zip/postal_code : 454171\n",
      " - contact_country : China\n",
      " - instrument_model : BGISEQ-500\n",
      " - library_selection : cDNA\n",
      " - library_source : transcriptomic\n",
      " - library_strategy : RNA-Seq\n",
      " - relation : BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN14207958, SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX7802549\n",
      " - supplementary_file_1 : ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4339nnn/GSM4339772/suppl/GSM4339772_C144_filtered_feature_bc_matrix.h5\n",
      " - series_id : GSE145926\n",
      " - data_row_count : 0\n",
      "Table data:\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "S2_geo_sample = GEOparse.get_GEO(geo=S2_geo_id, destdir=balf_dir)\n",
    "\n",
    "for key, value in S2_geo_sample.metadata.items():\n",
    "        print(\" - %s : %s\" % (key, \", \".join(value)))\n",
    "print (\"Table data:\",)\n",
    "print (S2_geo_sample.table.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse S2 raw HDF5 file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'> genome\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAACCTGAGCTAGTGG</th>\n",
       "      <th>AAACCTGCAGCCTTTC</th>\n",
       "      <th>AAACCTGCAGCTGTGC</th>\n",
       "      <th>AAACCTGGTCTAGTGT</th>\n",
       "      <th>AAACCTGGTGACTCAT</th>\n",
       "      <th>AAACCTGTCGTCCGTT</th>\n",
       "      <th>AAACGGGAGACGCTTT</th>\n",
       "      <th>AAACGGGAGATCGGGT</th>\n",
       "      <th>AAACGGGAGCGTCTAT</th>\n",
       "      <th>AAACGGGCAATCGAAA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL732372.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F29</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC114498.1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             AAACCTGAGCTAGTGG  AAACCTGCAGCCTTTC  AAACCTGCAGCTGTGC  \\\n",
       "MIR1302-2HG                 0                 0                 0   \n",
       "FAM138A                     0                 0                 0   \n",
       "OR4F5                       0                 0                 0   \n",
       "AL627309.1                  0                 0                 0   \n",
       "AL627309.3                  0                 0                 0   \n",
       "AL627309.2                  0                 0                 0   \n",
       "AL627309.4                  0                 0                 0   \n",
       "AL732372.1                  0                 0                 0   \n",
       "OR4F29                      0                 0                 0   \n",
       "AC114498.1                  0                 0                 0   \n",
       "\n",
       "             AAACCTGGTCTAGTGT  AAACCTGGTGACTCAT  AAACCTGTCGTCCGTT  \\\n",
       "MIR1302-2HG                 0                 0                 0   \n",
       "FAM138A                     0                 0                 0   \n",
       "OR4F5                       0                 0                 0   \n",
       "AL627309.1                  0                 0                 0   \n",
       "AL627309.3                  0                 0                 0   \n",
       "AL627309.2                  0                 0                 0   \n",
       "AL627309.4                  0                 0                 0   \n",
       "AL732372.1                  0                 0                 0   \n",
       "OR4F29                      0                 0                 0   \n",
       "AC114498.1                  0                 0                 0   \n",
       "\n",
       "             AAACGGGAGACGCTTT  AAACGGGAGATCGGGT  AAACGGGAGCGTCTAT  \\\n",
       "MIR1302-2HG                 0                 0                 0   \n",
       "FAM138A                     0                 0                 0   \n",
       "OR4F5                       0                 0                 0   \n",
       "AL627309.1                  0                 0                 0   \n",
       "AL627309.3                  0                 0                 0   \n",
       "AL627309.2                  0                 0                 0   \n",
       "AL627309.4                  0                 0                 0   \n",
       "AL732372.1                  0                 0                 0   \n",
       "OR4F29                      0                 0                 0   \n",
       "AC114498.1                  0                 0                 0   \n",
       "\n",
       "             AAACGGGCAATCGAAA  \n",
       "MIR1302-2HG                 0  \n",
       "FAM138A                     0  \n",
       "OR4F5                       0  \n",
       "AL627309.1                  0  \n",
       "AL627309.3                  0  \n",
       "AL627309.2                  0  \n",
       "AL627309.4                  0  \n",
       "AL732372.1                  0  \n",
       "OR4F29                      0  \n",
       "AC114498.1                  0  "
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s2_raw = get_matrix_from_h5(os.path.join(balf_dir,S2_raw_file))\n",
    "#print(s2_raw.iloc[:10,:10].to_html())\n",
    "s2_raw.iloc[:10,:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33539, 3716)"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s2_raw.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare Raw S2 barcodes with S2 metadata barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3716, 16883)"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s2_metadata_barcodes = set(s2_cells['ID'].str.split('_').str.get(0).values.tolist())\n",
    "s2_raw_barcodes = set([raw.split('-')[0] for raw in s2_raw.columns])\n",
    "len(s2_raw_barcodes), len(s2_metadata_barcodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "96"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(s2_raw_barcodes) & set(s2_metadata_barcodes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(309, 4)"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s2_epithelial_cells = s2_cells[s2_cells['celltype'] == 'Epithelial']\n",
    "s2_epithelial_metadata_barcodes = set(s2_epithelial_cells['ID'].str.split('_').str.get(0).values.tolist())\n",
    "len(s2_epithelial_metadata_barcodes), len(s2_raw_barcodes & s2_epithelial_metadata_barcodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Healthy controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>group</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>celltype</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th>sample_new</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C100</th>\n",
       "      <th>HC3</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C51</th>\n",
       "      <th>HC1</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C52</th>\n",
       "      <th>HC2</th>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3660650</th>\n",
       "      <th>HC4</th>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "      <td>1383</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         ID  group  disease  hasnCoV  cluster  celltype  \\\n",
       "sample     sample_new                                                     \n",
       "C100       HC3           13     13       13       13       13        13   \n",
       "C51        HC1           13     13       13       13       13        13   \n",
       "C52        HC2           33     33       33       33       33        33   \n",
       "GSM3660650 HC4         1383   1383     1383     1383     1383      1383   \n",
       "\n",
       "                       barcode  \n",
       "sample     sample_new           \n",
       "C100       HC3              13  \n",
       "C51        HC1              13  \n",
       "C52        HC2              33  \n",
       "GSM3660650 HC4            1383  "
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "healthy_controls = all_cells[(all_cells['group'] == 'HC') & (all_cells['celltype'] == 'Epithelial')].groupby(['sample', 'sample_new']).count()\n",
    "healthy_controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "patients = all_cells.groupby('sample_new').count().index.tolist()\n",
    "patients_no_HC4 = [p for p in patients if p != 'HC4']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write out epithelial samples to csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "for patient in patients_no_HC4:\n",
    "    raw[patient].columns = [barcode.split('-')[0] for barcode in raw[patient].columns]\n",
    "    #raw[patient].to_csv(os.path.join(\n",
    "    #    balf_dir,\n",
    "    #    '{}_umi_full.csv'.format(patient)))\n",
    "    #raw[patient][metadata_barcodes[patient]].to_csv(\n",
    "    #    os.path.join(balf_dir, 'Annotated',\n",
    "    #                 '{}_umi_annotated.csv'.format(patient)))\n",
    "    raw[patient][epithelial_barcodes[patient]].to_csv(\n",
    "        os.path.join(balf_dir, 'Epithelial',\n",
    "                    '{}_umi_epithelial.csv'.format(patient)))\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmgpp = {}\n",
    "for patient in patients_no_HC4:\n",
    "    tmgpp[patient] = raw[patient].loc[model_genes].sum(axis=1)\n",
    "tmgpp['HC4'] = hc4_raw.loc[model_genes].sum(axis=1)\n",
    "total_model_genes_per_patient = pd.DataFrame(tmgpp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HC1</th>\n",
       "      <th>HC2</th>\n",
       "      <th>HC3</th>\n",
       "      <th>HC4</th>\n",
       "      <th>M1</th>\n",
       "      <th>M2</th>\n",
       "      <th>M3</th>\n",
       "      <th>S1</th>\n",
       "      <th>S2</th>\n",
       "      <th>S3</th>\n",
       "      <th>S4</th>\n",
       "      <th>S5</th>\n",
       "      <th>S6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IL6</th>\n",
       "      <td>41</td>\n",
       "      <td>7</td>\n",
       "      <td>57</td>\n",
       "      <td>89.0</td>\n",
       "      <td>2783</td>\n",
       "      <td>1293</td>\n",
       "      <td>18</td>\n",
       "      <td>2773</td>\n",
       "      <td>3530</td>\n",
       "      <td>101</td>\n",
       "      <td>677</td>\n",
       "      <td>851</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STAT3</th>\n",
       "      <td>3914</td>\n",
       "      <td>1739</td>\n",
       "      <td>10790</td>\n",
       "      <td>2772.0</td>\n",
       "      <td>9554</td>\n",
       "      <td>8607</td>\n",
       "      <td>786</td>\n",
       "      <td>8478</td>\n",
       "      <td>25927</td>\n",
       "      <td>1995</td>\n",
       "      <td>4044</td>\n",
       "      <td>2663</td>\n",
       "      <td>4275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACE2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>45.0</td>\n",
       "      <td>56</td>\n",
       "      <td>24</td>\n",
       "      <td>12</td>\n",
       "      <td>45</td>\n",
       "      <td>59</td>\n",
       "      <td>21</td>\n",
       "      <td>305</td>\n",
       "      <td>138</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADAM17</th>\n",
       "      <td>11892</td>\n",
       "      <td>4586</td>\n",
       "      <td>21162</td>\n",
       "      <td>880.0</td>\n",
       "      <td>12249</td>\n",
       "      <td>10612</td>\n",
       "      <td>2172</td>\n",
       "      <td>11119</td>\n",
       "      <td>25206</td>\n",
       "      <td>1763</td>\n",
       "      <td>4799</td>\n",
       "      <td>1691</td>\n",
       "      <td>3566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGTR1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGF</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6</td>\n",
       "      <td>17</td>\n",
       "      <td>86</td>\n",
       "      <td>34</td>\n",
       "      <td>28</td>\n",
       "      <td>3</td>\n",
       "      <td>37</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IL6R</th>\n",
       "      <td>1268</td>\n",
       "      <td>920</td>\n",
       "      <td>2838</td>\n",
       "      <td>379.0</td>\n",
       "      <td>7255</td>\n",
       "      <td>7394</td>\n",
       "      <td>522</td>\n",
       "      <td>8649</td>\n",
       "      <td>25538</td>\n",
       "      <td>1897</td>\n",
       "      <td>3561</td>\n",
       "      <td>910</td>\n",
       "      <td>3554</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             HC1   HC2    HC3     HC4     M1     M2    M3     S1     S2    S3  \\\n",
       "gene_name                                                                       \n",
       "IL6           41     7     57    89.0   2783   1293    18   2773   3530   101   \n",
       "STAT3       3914  1739  10790  2772.0   9554   8607   786   8478  25927  1995   \n",
       "ACE2           2     0      1    45.0     56     24    12     45     59    21   \n",
       "ADAM17     11892  4586  21162   880.0  12249  10612  2172  11119  25206  1763   \n",
       "AGTR1          0     0      0     3.0      1      0     0      0      0     0   \n",
       "EGF            0     0      1     5.0      6     17    86     34     28     3   \n",
       "IL6R        1268   920   2838   379.0   7255   7394   522   8649  25538  1897   \n",
       "\n",
       "             S4    S5    S6  \n",
       "gene_name                    \n",
       "IL6         677   851   115  \n",
       "STAT3      4044  2663  4275  \n",
       "ACE2        305   138    35  \n",
       "ADAM17     4799  1691  3566  \n",
       "AGTR1         0     0     0  \n",
       "EGF          37     5     8  \n",
       "IL6R       3561   910  3554  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "total_model_genes_per_patient[sorted(total_model_genes_per_patient.columns)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": false,
   "autoclose": false,
   "autocomplete": false,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": true,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
